```{r}
library(vroom)
library(tidyverse)
library(qvalue)
library(RColorBrewer)
```

First, do the CellRegMap eQTLs capture the dynamic eQTLs that we detected?

Load relevant summary stats
```{r}
cm_emt <- vroom("results/dynamic_qtl_calling/eb-cm_15binstrimmed/pseudobulk_tmm/nipals/10clpcs/tensorqtl_interactions.cis_qtl_top_assoc.txt.gz")
cm_hits <- filter(cm_emt, pval_adj_bh <= 0.1)

neur_emt <- vroom("results/dynamic_qtl_calling/eb-neur_15binstrimmed/pseudobulk_tmm/nipals/10clpcs/tensorqtl_interactions.cis_qtl_top_assoc.txt.gz")
neur_hits <- filter(neur_emt, pval_adj_bh <= 0.1)

hep_emt <- vroom("results/dynamic_qtl_calling/eb-hep_15binstrimmed/pseudobulk_tmm/nipals/10clpcs/tensorqtl_interactions.cis_qtl_top_assoc.txt.gz")
hep_hits <- filter(hep_emt, pval_adj_bh <= 0.1)

crm <- vroom("results/cellregmap_eqtl_calling/eb_cellid/pseudobulk_tmm/basic/all_genes_merged.mash-signif.fasttopics_10topics.cellregmap.tsv")
```

Check each trajectory
```{r}
cm_rep <- inner_join(crm, cm_hits, by=c("EB_HGNC"="phenotype_id", "EB_VARIANT_ID"="variant_id"))
neur_rep <- inner_join(crm, neur_hits, by=c("EB_HGNC"="phenotype_id", "EB_VARIANT_ID"="variant_id"))
hep_rep <- inner_join(crm, hep_hits, by=c("EB_HGNC"="phenotype_id", "EB_VARIANT_ID"="variant_id"))
```

```{r}
hist(cm_rep$P_CELLREGMAP)
hist(neur_rep$P_CELLREGMAP)
hist(hep_rep$P_CELLREGMAP)
```

We don't have the power to make any strong claims about replication rates
Keep in mind we only tested significant mash eQTLs in CRM, so not surprising

### Visualizing CRM
Load all effect size estimates of significant CRM eQTLs
```{r}
crm_sighits <- vroom("results/cellregmap_eqtl_calling/eb_cellid/pseudobulk_tmm/basic/all_genes_merged.mash-signif.fasttopics_10topics.cellregmap.sighits.tsv")
crm_siggenes <- crm_sighits$EB_HGNC

crm_betas <- vroom(paste0("results/cellregmap_eqtl_calling/eb_cellid/pseudobulk_tmm/basic/mash-signif.fasttopics_10topics.",crm_siggenes, ".cellregmap.betas.tsv"))

crm_betas_wide <- select(crm_betas, c(PSEUDOCELL, BETA_GXC, EB_HGNC, EB_VARIANT_ID)) %>%
  unite(EB_QTL, EB_HGNC, EB_VARIANT_ID, sep="_") %>%
  pivot_wider(id_cols=PSEUDOCELL, names_from=EB_QTL, values_from=BETA_GXC)
```

Next, load the topic loadings
```{r}
topic_loadings <- vroom("results/fast_topics/fasttopics_10topics_loadings.tsv")

crm_betas_loadings <- left_join(crm_betas_wide, topic_loadings, by=c("PSEUDOCELL"="pseudocell"))

betas_only <- crm_betas_loadings[, seq(ncol(crm_betas_loadings)-10)] %>%
  column_to_rownames("PSEUDOCELL")
topics_only <- crm_betas_loadings[, c(1, seq(ncol(crm_betas_loadings)-9, ncol(crm_betas_loadings)))] %>%
  column_to_rownames("PSEUDOCELL")

beta_topic_cor <- cor(betas_only, topics_only)

write_tsv(as_tibble(betas_only, rownames="PSEUDOCELL"), "results/cellregmap_eqtl_calling/eb_cellid/pseudobulk_tmm/basic/all_genes_merged.mash-signif.fasttopics_10topics.cellregmap.betas.tsv")
```

Use this cor matrix to prioritize effects for visualization
```{r}
qtl_prioritization <- as_tibble(beta_topic_cor, rownames="qtl") %>%
  separate(qtl, into=c("gene", "snp"), sep="_")

qtl_candidates <- filter(qtl_prioritization, gene %in% neur_exp$gene)
```


One that may be a nonlinear dynamic eQTL is GSAP
```{r}
neur_exp <- vroom("data/dynamic_qtl_calling/eb-neur_15binstrimmed/pseudobulk_tmm/expression.bed.gz") %>%
  filter(gene %in% qtl_candidates$gene) %>%
  select(-c(`#chr`, start, end)) %>%
  pivot_longer(!gene, names_to="ind_type", values_to="expression")
neur_genotype_df <- vroom("data/dynamic_qtl_calling/eb-neur_15binstrimmed/pseudobulk_tmm/5clpcs/genotype_df.tsv") %>% 
  filter(snp %in% qtl_candidates$snp) %>%
  pivot_longer(!snp, names_to="ind_type", values_to="genotype")
neur_pseudotime <- vroom("data/dynamic_qtl_calling/eb-neur_15binstrimmed/pseudobulk_tmm/pseudotime.tsv") %>%
  mutate(pseudotime_bin=as.numeric(str_extract(ind_type, "[^bin]+$")))
```

```{r}
g = "CNTN3"
snp = "rs7433110"

select(filter(neur_exp, gene==!!g), c("ind_type", "expression")) %>%
  inner_join(filter(neur_genotype_df, snp==!!snp), c("ind_type", "expression"), by="ind_type") %>%
  inner_join(neur_pseudotime, by="ind_type") %>%
  mutate(genotype=factor(genotype), pseudotime_bin=factor(pseudotime_bin)) %>%
  ggplot(aes(x=pseudotime_bin, y=expression, fill=genotype)) +
  geom_boxplot() +
  theme_classic()
```
```{r}
g = "CEP78"
snp = "rs10867166"

select(filter(neur_exp, gene==!!g), c("ind_type", "expression")) %>%
  inner_join(filter(neur_genotype_df, snp==!!snp), c("ind_type", "expression"), by="ind_type") %>%
  inner_join(neur_pseudotime, by="ind_type") %>%
  mutate(genotype=factor(genotype), pseudotime_bin=factor(pseudotime_bin)) %>%
  ggplot(aes(x=pseudotime_bin, y=expression, fill=genotype)) +
  geom_boxplot() +
  theme_classic()
```

```{r}
g = "PPIP5K2"
snp = "rs34829"

select(filter(neur_exp, gene==!!g), c("ind_type", "expression")) %>%
  inner_join(filter(neur_genotype_df, snp==!!snp), c("ind_type", "expression"), by="ind_type") %>%
  inner_join(neur_pseudotime, by="ind_type") %>%
  mutate(genotype=factor(genotype), pseudotime_bin=factor(pseudotime_bin)) %>%
  ggplot(aes(x=pseudotime_bin, y=expression, fill=genotype)) +
  geom_boxplot() +
  theme_classic()
```

```{r}
g = "AS3MT"
snp = "rs11191419"

select(filter(neur_exp, gene==!!g), c("ind_type", "expression")) %>%
  inner_join(filter(neur_genotype_df, snp==!!snp), c("ind_type", "expression"), by="ind_type") %>%
  inner_join(neur_pseudotime, by="ind_type") %>%
  mutate(genotype=factor(genotype), pseudotime_bin=factor(pseudotime_bin)) %>%
  ggplot(aes(x=pseudotime_bin, y=expression, fill=genotype)) +
  geom_boxplot(lwd=1.25) +
  theme_classic(base_size=20) +
  scale_fill_manual(values=brewer.pal(3, "Set1")) +
  xlab("Pseudotime Bin") +
  ylab("AS3MT Expression") +
  labs(fill="Genotype at rs11191419")
```

Look at the structure plot for this variant
```{r}
# topics_kept <- paste0("k", seq(1, 10))
topics_kept <- c('k1','k2', 'k4','k5','k6', 'k7', 'k8', 'k10')
structure_df <- crm_betas_loadings %>%
  select(all_of(c("PSEUDOCELL", "AS3MT_rs11191419", topics_kept))) %>%
  pivot_longer(-c(PSEUDOCELL, `AS3MT_rs11191419`), names_to="topic", values_to="loading") %>%
  mutate(topic=factor(topic, levels=topics_kept))
```

```{r, fig.width=8, fig.height=4}
ggplot(structure_df, aes(x=AS3MT_rs11191419, y=loading, fill=topic)) +
  geom_col(width=1e-3) +
  theme_classic(base_size=35) +
  scale_fill_manual(values=brewer.pal(10, 'Set1')) +
  theme(legend.position="none") +
  xlab("Aggregate Interacting Environment at rs11191419") +
  ylab("Topic Loadings")
```

Also get the pseudobulk expression 
```{r}
pseudocell_exp <- vroom("data/single_cell_objects/eb_pseudocells_normalized.tsv")

expression_df <- crm_betas_loadings %>%
  select(all_of(c("PSEUDOCELL", "AS3MT_rs11191419"))) %>% 
  left_join(select(pseudocell_exp, c(`...1`, AS3MT)), by=c("PSEUDOCELL"="...1"))
```

