---
title: "Distribution of eQTLs around developmental genes"
output: html_notebook
---

```{r}
library(tidyverse)
library(Matrix.utils)
library(fgsea)
library(susieR)
library(pathways)
library(vroom)
library(gseasusie)
set.seed(1)
```

```{r}
eb_sighits_bed_loc <- "/project2/gilad/jpopp/ebQTL/results/static_eqtl_followup/qtl_sets/tensorqtl/original/signif_variant_gene_pairs.bed"
eb_sighits_overlap_loc <- "/project2/gilad/jpopp/ebQTL/results/static_eqtl_followup/qtl_sets/tensorqtl/original/signif_variant_gene_pairs.all_tissue_overlap.bed"
all_tests_loc <- "results/static_qtl_calling/eb_cellid/pseudobulk_tmm/basic/8pcs/tensorqtl_nominal.all.tsv"
gmt_loc <- "data/gene_sets/c5.go.bp.v2022.1.Hs.symbols.gmt"
```

Do our novel eQTLs display the anticipated properties of eQTLs, or could they just be
a result of the depletion of GTEx eQTLs for these gene sets?

First, load the gene sets
```{r}
gmt_lines <- readLines(gmt_loc)

gmt_list <- lapply(gmt_lines, function(x) unlist(strsplit(x, "\t")))
gmt_list2 <- lapply(gmt_list, function(x) list(x[1], x[2], x[3:length(x)]))

gmt_df <- as_tibble(do.call("rbind", gmt_list2)) %>%
  mutate(across(V1:V2, unlist)) %>%
  dplyr::rename(geneset=V1, link=V2, genes=V3) %>%
  filter(str_sub(geneset, 1, 4) == "GOBP") # filter to BP 

gmt_mat <- gmt_df %>%
  dplyr::select(-c(link)) %>%
  unnest(genes) %>%
  filter(genes %in% eb_distinct$EB_HGNC) %>%
  pivot_wider(names_from = genes, values_from = genes, values_fn = length, values_fill = 0) %>%
  column_to_rownames("geneset")

tissue_dev_genes <- rownames(gmt_mat)[which(gmt_mat[,"GOBP_TISSUE_DEVELOPMENT"]==1)]
```

Next, load the variant-gene pairs we'll look at
```{r}
all_tests <- vroom(all_tests_loc)

eb_distinct <- vroom(eb_sighits_bed_loc) %>% 
  dplyr::select(c(EB_HGNC, EB_VARIANT_ID)) %>%
  distinct()

gtex_distinct <- vroom(eb_sighits_overlap_loc, col_names=c("CHR", "START", "STOP", "EB_ENSG", "EB_HGNC", 
                                                          "RSID", "CELLTYPE", "CHR1", "START1", 
                                                          "STOP1", "GTEX_ENSG", "GTEX_REF", "GTEX_ALT")) %>%
  dplyr::select(EB_HGNC, RSID) %>%
  distinct()

eb_novel <- anti_join(eb_distinct, gtex_distinct, by=c("EB_HGNC"="EB_HGNC", "EB_VARIANT_ID"="RSID"))
```

Tidy up the three sets of variants we'll be looking at
```{r}
tested_variants_dev <- filter(all_tests, phenotype_id %in% tissue_dev_genes) %>%
  select(phenotype_id, variant_id) %>%
  mutate(group="All tested variants")
qtl_variants_dev <- filter(eb_distinct, EB_HGNC %in% tissue_dev_genes) %>%
  dplyr::rename(phenotype_id=EB_HGNC, variant_id=EB_VARIANT_ID) %>%
  mutate(group="HDC eQTL variants")
novel_variants_dev <- anti_join(qtl_variants_dev, gtex_distinct, by=c("phenotype_id"="EB_HGNC", "variant_id"="RSID")) %>%
  mutate(group="Novel HDC eqtl variants")

```


Now, load a few properties we'll need: gene TSS and variant position
```{r}
gtf <- vroom("data/gencode/gencode.hg38.filtered.gtf") %>%
  mutate(tss = if_else(strand == "+", start, end)) %>%
  dplyr::select(hgnc, tss)

bim <- vroom("data/static_qtl_calling/eb_cellid/pseudobulk_tmm/basic/all_celltypes_combined/genotypes_filtered_plink.bim",
             col_names=c("#CHR", "variant_id", "POS_CM", "POS_BP", "ALLELE_1", "ALLELE_2"),
             col_select=c("#CHR", "variant_id", "POS_BP"))

tss_comparison <- bind_rows(tested_variants_dev,
                            qtl_variants_dev,
                            novel_variants_dev) %>%
  left_join(gtf, by=c("phenotype_id"="hgnc")) %>%
  left_join(select(bim, c(variant_id, POS_BP)), by="variant_id") %>%
  mutate(tssdist=map2_int(tss, POS_BP, `-`))
```

And compare the distribution of the three variant groups
```{r}
tss_comparison <- mutate(tss_comparison, group=factor(group, levels=c("All tested variants",
                                                                      "HDC eQTL variants",
                                                                      "Novel HDC eqtl variants")))
  
ggplot(tss_comparison, aes(x=tssdist, color=group)) +
  geom_density(linewidth=1.1) +
  theme_classic(base_size=15) +
  xlab("Distance to TSS") + ylab("Density") +
  labs(color="Variant Class") +
  scale_color_manual(values=c("#D5D5D5", "#004D80", "#D41876")) +
  guides(colour = guide_legend(override.aes = list(shape = 10)))
```
