```{r}
library(tidyverse)
library(vroom)
library(biomaRt)
library(data.table)
library(patchwork)
```


```{r}
alm_gwas <- vroom("data/gwas/pei/GCST90000025.sumstats.hg38.bed",
                  col_names=c("#CHR", "START", "END", "HG19", "P")) %>%
  filter(`#CHR` == "7")

cm_dqtl <- vroom("results/dynamic_qtl_calling/eb-cm_15binstrimmed/pseudobulk_tmm/nipals/10clpcs/tensorqtl_interactions.all.tsv")
col1a2_qtl <- filter(cm_dqtl, phenotype_id == "COL1A2")

cm_dqtl_bim <- vroom("data/dynamic_qtl_calling/eb-cm_15binstrimmed/pseudobulk_tmm/nipals/genotypes_filtered_plink.bim",
                      col_names=c("CHR", "RS", "NA", "POS", "REF", "ALT")) %>%
  filter(CHR == "chr7") %>%
  dplyr::select(RS, POS)
```


```{r}
col1a2_tss <- 94394561
alm_gwas_merge <- alm_gwas %>%
  dplyr::select(END, P) %>%
  filter((col1a2_tss - 50e3 <= END) & (col1a2_tss + 50e3 >= END)) %>%
  mutate(assay="GWAS")

col1a2_qtl_merge <- col1a2_qtl %>%
  dplyr::select(variant_id, pval_gi) %>%
  inner_join(cm_dqtl_bim, by=c("variant_id"="RS")) %>%
  dplyr::select(-c(variant_id)) %>%
  dplyr::rename(END=POS, P=pval_gi) %>%
  mutate(assay="QTL") 
  
locus_view <- bind_rows(alm_gwas_merge, col1a2_qtl_merge) %>%
  mutate(nlog10p=-log10(P))
```

Are the most significant hits really overlapping? Hard to tell here
```{r}
gwas_strongest <- filter(locus_view, assay == "GWAS") %>%
  filter(nlog10p >= 15) %>%
  pull(END)
locus_view %>%
  mutate(gwas_hit=END %in% gwas_strongest) %>%
  ggplot(aes(x=END, y=nlog10p, color=gwas_hit)) +
  facet_grid(rows=vars(assay), scales="free_y") +
  geom_point() +
  geom_vline(aes(xintercept=col1a2_tss), color="red") +
  theme_classic() +
  scale_color_manual(values=c("black", "red"))
```

## Colored by LD
Let's view the LD patterns underlying the primary locus, to see if it suggests this is one or multiple loci
```{r}
# gwas_vars <- alm_gwas_merge %>%
#   dplyr::select(END, P) %>% 
#   rename(POS=END) %>%
#   mutate(CHR="chr7", .before=1) %>%
#   mutate(SNP=paste(CHR, POS, sep=":"))
# 
# ldmat <- LDlinkR::LDmatrix(gwas_vars$SNP, 
#                   pop = "EUR", 
#                   r2d = "r2",
#                   token = "36e356ef3c1e", 
#                   file = FALSE,
#                   genome_build = "grch38")
# write_tsv(ldmat, "temp/ld_matrix.txt")
```

Map these RS IDs to chromosome positions
```{r}
ldmat <- read.table("temp/ld_matrix.txt", header=T)

snpmart = useEnsembl(biomart = "snps", dataset="hsapiens_snp")
snp_list <- getBM(attributes = c('refsnp_id','allele','chrom_start','chrom_strand'), 
      filters = c('snp_filter'), 
      values = ldmat$RS_number, 
      mart = snpmart)
snp_pos <- as_tibble(snp_list) 

lead_variant_1 <- filter(snp_pos, chrom_start==94346257)$refsnp_id
ldmat_pos <- as_tibble(ldmat) %>%
  left_join(dplyr::select(snp_pos, -c(chrom_strand)), by=c("RS_number"="refsnp_id")) %>%
  dplyr::select(c(RS_number, chrom_start, lead_variant_1))
```

Plot the GWAS with correlation to the first lead SNP
```{r}
locus_view_lv1_ld <- left_join(locus_view, dplyr::select(ldmat_pos, c(chrom_start, lead_variant_1)), by=c("END"="chrom_start"))

locus_view_lv1_ld %>%
  filter(assay == "GWAS") %>%
  ggplot(aes(x=END, y=nlog10p, color=rs55921499)) +
      facet_grid(rows=vars(assay), scales="free_y") +
      geom_point() +
      geom_vline(aes(xintercept=col1a2_tss), color="red") +
      theme_classic() +
      scale_color_distiller(palette = "Spectral")
```
Next, we need LD on the QTL side from our YRI donors

List SNPs included in the QTL plot
```{r}
rs_col1a2 <- filter(cm_dqtl_bim, POS %in% col1a2_qtl_merge$END)
```

Get the genotype data
```{r}
cm_genotype_df_full <- vroom("data/dynamic_qtl_calling/eb-cm_15binstrimmed/pseudobulk_tmm/nipals/10clpcs/genotype_df.tsv") 
cm_genotype_df_col1a2 <- cm_genotype_df_full %>% 
  filter(snp %in% rs_col1a2$RS) %>%
  column_to_rownames("snp")
ld_col1a2 <- cor(t(cm_genotype_df_col1a2))^2
```

Get the nearest tag variant in the GWAS that we tested in EBs
```{r}
strongest_tag_snp <- dplyr::select(ldmat, c(RS_number,rs55921499)) %>%
  filter(RS_number %in% rs_col1a2$RS) %>%
  arrange(desc(rs55921499)) %>%
  slice_head(n=1) %>%
  pull(RS_number)
```

Get LD with the lead variant in YRI subpopulation
```{r}
col1a2_qtl_ld <- as_tibble(ld_col1a2[,strongest_tag_snp], rownames="RS") %>%
  inner_join(rs_col1a2, by="RS") %>%
  right_join(col1a2_qtl_merge, by=c("POS"="END"))

ggplot(col1a2_qtl_ld, aes(x=POS, y=-log10(P), color=value)) +
  geom_point() +
  scale_color_distiller(palette = "Spectral")
```

Combined the two plots colored by LD
```{r, fig.height=6, fig.width=6}
alm_ld_plot <- locus_view_lv1_ld %>%
  filter(assay=="GWAS") %>%
  dplyr::select(c(END, rs55921499, nlog10p, assay)) %>%
  rename(lead_ld=rs55921499, POS=END)

p1 <- ggplot(alm_ld_plot, aes(x=POS, y=nlog10p, color=lead_ld)) +
  geom_point(data=filter(alm_ld_plot, lead_ld != 1), size=3) +
  scale_color_distiller(palette = "Spectral") +
  geom_point(data=filter(alm_ld_plot, lead_ld == 1), color="purple", fill="black", shape=23, size=3.5, stroke=2) +
  geom_point(data=filter(alm_ld_plot, POS == 94374529), color="purple", fill="black", shape=22, size=3.5, stroke=2) +
  theme_classic(base_size=20) +
  ylab("-log10(P)") + xlab("Position (chr7)") +
  labs(color="LD with lead variant,\nrs55921499 (EUR)") +
  ggtitle("ALM GWAS")
  
col1a2_ld_plot <- col1a2_qtl_ld %>%
  mutate(nlog10p = -log10(P)) %>%
  dplyr::select(c(RS, POS, value, nlog10p)) %>%
  rename(tag_ld=value) %>%
  mutate(assay="QTL")

p2 <- ggplot(col1a2_ld_plot, aes(x=POS, y=nlog10p, color=tag_ld)) +
  geom_point(size=3) +
  scale_color_distiller(palette = "Spectral") +
  geom_point(data=filter(col1a2_ld_plot, RS == strongest_tag_snp), color="purple", fill="black", shape=22, size=3.5, stroke=2) +
  theme_classic(base_size=20) +
  ylab("-log10(P)") + xlab("Position (chr7)") +
  labs(color="LD with strongest tag variant\nassayed, rs17166205\n(YRI subset)") +
  ggtitle("COL1A2 QTL")

p1 / p2
```

