---
title: "Enrichment of eQTLs for schizophrenia risk"
output: html_notebook
---
```{r}
library(vroom)
library(data.table)
library(tidyverse)
library(snpStats)
library(RColorBrewer)
set.seed(42)
```


### Disease
Load eQTLs
```{r}
tensorqtl_eqtls <- vroom("results/static_eqtl_followup/qtl_sets/tensorqtl/original/signif_variant_gene_pairs.bed")
```


```{r}
dbp <- fread("data/gwas/keaton_2024/keaton-dbp.sumstats.hg38.bed")

dbp_eqtls <- inner_join(mutate(tensorqtl_eqtls, `#CHR`=str_replace(`#CHR`, "chr", "")), dbp, by=c("#CHR"="V1", "END"="V3"))
random_snps_p <- sort(sample(dbp$V5, nrow(dbp_eqtls)))

qq_dbp <- tibble(eqtls=dbp_eqtls$V5) %>%
  arrange(eqtls) %>%
  rowid_to_column() %>%
  mutate(eqtls=-log10(eqtls)) %>%
  mutate(bg=-log10(random_snps_p)) %>%
  mutate(null = -log10(rowid / nrow(.) )) %>%
  pivot_longer(cols=c(eqtls, bg), names_to="variant", values_to="p") %>%
  mutate(variant=factor(variant, levels=c("bg", "eqtls")))

dbp_qq_plot <- ggplot(qq_dbp, aes(x=null, y=p, color=variant)) +
  geom_point() + 
  geom_abline(slope=1, intercept=0, linetype="dashed") +
  theme_classic(base_size=20) +
  xlab("Theoretical -log10(P)") + ylab("Empirical -log10(P)") +
  scale_color_manual(values=c("#929292", "#B51700"), labels=c("Random Variants", "HDC eQTLs")) +
  labs(color="Variant Group") +
  #theme(legend.position="none") +
  ggtitle("Diastolic Blood Pressure")
```

## How many of these dbp eQTLs are significant?
```{r}
dbp_eqtl_hits <- filter(dbp_eqtls, V5 <= 5e-8)
```

How many of these dbp eQTLs are novel?
```{r}
dbp_eqtl_novel_hits_bed <- vroom("results/static_eqtl_followup/qtl_sets/tensorqtl/dbp_overlap_keaton/signif_variant_gene_pairs.gtex_removed.tagged_gtex_removed.bed")

length(unique(dbp_eqtl_hits$EB_HGNC))
length(unique(dbp_eqtl_novel_hits_bed$EB_HGNC))
```

## Novel eQTL for BCL11B
```{r}
snpstats2geno <- function(s) {
  ac <- tibble(ac=as.numeric(s$genotypes@.Data), sample=rownames(s$genotypes@.Data))
  ac2geno <- tibble("ac"=c(1, 2, 3), 
                    "genotype"=c(paste(rep(s$map$allele.1, 2), sep="", collapse=""), 
                                                  paste(s$map$allele.1, s$map$allele.2, sep="", collapse=""), 
                                                  paste(rep(s$map$allele.2, 2), sep="", collapse="")))
  geno <- left_join(ac, ac2geno, by="ac")
  geno
}
```


```{r}
snp_rsid <- "rs4763297"
gene_hgnc <- "APOLD1"
celltype <- "Stromal-cells"

bed <- paste0("/project2/gilad/jpopp/ebQTL/data/static_qtl_calling/eb_cellid/pseudobulk_tmm/basic/", celltype, "/genotypes_filtered_plink.bed")
bim <- paste0("/project2/gilad/jpopp/ebQTL/data/static_qtl_calling/eb_cellid/pseudobulk_tmm/basic/", celltype, "/genotypes_filtered_plink.bim")
fam <- paste0("/project2/gilad/jpopp/ebQTL/data/static_qtl_calling/eb_cellid/pseudobulk_tmm/basic/", celltype, "/genotypes_filtered_plink.fam")
snpstat <- read.plink(bed, bim, fam, na.strings = c("0", "-9"), sep = "." , select.subjects = NULL, select.snps = snp_rsid)

genotypes_df <- snpstats2geno(snpstat)

expression_df <- vroom(paste0("/project2/gilad/jpopp/ebQTL/data/static_qtl_calling/eb_cellid/pseudobulk_tmm/basic/", celltype, "/expression.bed.gz")) %>%
  select(-c(`#chr`, start, end)) %>%
  column_to_rownames("gene") %>% t %>%
  as_tibble(rownames="sample") %>%
  dplyr::select(c(sample, gene_hgnc))
```

```{r}
dbp_qtl_plot <- inner_join(genotypes_df, expression_df, by="sample") %>%
  ggplot(aes(x=genotype, y=.data[[gene_hgnc]], color=genotype)) +
  geom_boxplot(alpha=0.25, width=0.5, lwd=1.5) +
  geom_jitter(aes(fill=genotype), position = position_jitter(seed = 1, width = 0.1), size=3.5, shape=21, lwd=1.25, color="black") +
  theme_classic(base_size=20) +
  scale_fill_manual(name=snp_rsid, values=brewer.pal(3, "Set1")) +
  #xlab("Genotype") + ylab(paste0(gene_hgnc, " Expression")) +
  xlab(snp_rsid) + ylab(paste0(gene_hgnc, " Expression\n(", celltype, ")")) +
  scale_color_manual(values=brewer.pal(3, "Set1")) +
  theme(legend.position="none")

save(dbp_qtl_plot, file="/project2/gilad/jpopp/ebQTL/temp/dbp_qtl_plot.Rdata")
```
