---
title: "Cell-ID Evaluation (EB Side)"
output: html_notebook
---

```{r}
library(tidyverse)
library(Matrix.utils)
library(fgsea)
library(susieR)
library(pathways)
library(vroom)
library(gseasusie)
library(scater)
library(Nebulosa)
```

```{r}
### Load & tidy data
# Read in the input files from Snakefile
eb_loc <- snakemake@input[["fca"]]
celltype_signatures <- snakemake@input[["signatures"]]
table_prefix <- snakemake@params[['table_prefix']]
```

Or
```{r}
eb_loc <- "/project2/gilad/jpopp/ebQTL/data/single_cell_objects/highpass/eb_pflog1ppfnorm.fca_hvgs.mca.sce"
load("/project2/gilad/jpopp/ebQTL/data/fca/cellid_signatures_and_hvgs.Rdata")
```

Load cell and gene embeddings
```{r}
eb <- readRDS(eb_loc)
eb$celltype[eb$celltype=="unassigned"] <- NA
# Get a list of cell types
cell.types <- unique(eb$celltype)
cell.types <- cell.types[!is.na(cell.types)]

# Pull MCA embeddings of cells and genes
cell_emb <- as_tibble(reducedDim(eb, "MCA"), rownames="cell") %>%
  left_join(as_tibble(colData(eb), rownames="cell"), by="cell") 
gene_emb <- attributes(cell_emb$MCA_1)$genesCoordinates

# Get cell type centroids
celltype_centroids <- as.matrix(aggregate.Matrix(reducedDim(eb, "MCA"), groupings=eb$celltype, fun="mean"))
```


# Driver gene sets for problem annotations 
## Immune Cells
For lymphoid and megakaryocyte cell types, find 100 closest genes
```{r}
find_closest <- function(t, num.neighbors=100) {
  centroid_t <- t(celltype_centroids[t,])
  diffs_t <- sweep(gene_emb, 2, centroid_t)
  markers_t <- apply(diffs_t, 1, norm, type="2") %>%
    as_tibble(rownames="gene") %>%
    arrange(value) %>% slice_head(n=num.neighbors) %>%
    pull(gene)
  return(markers_t)
}

lymphoid_closest_genes <- lapply(c("Lymphoid cells"), find_closest)
lymphoid_drivers <- intersect(lymphoid_closest_genes[[1]], new_signatures[['Lymphoid cells']])

megakaryocyte_closest_genes <- lapply(c("Megakaryocytes"), find_closest)
megakaryocyte_drivers <- intersect(megakaryocyte_closest_genes[[1]], new_signatures[['Megakaryocytes']])
```

## Epithelial Cells
### Acinar cells
```{r}
acinar_closest_genes <- lapply(c("Acinar cells"), find_closest)
acinar_drivers <- intersect(acinar_closest_genes[[1]], new_signatures[['Acinar cells']])
```

### Ductal cells
```{r}
ductal_closest_genes <- lapply(c("Ductal cells"), find_closest)
ductal_drivers <- intersect(ductal_closest_genes[[1]], new_signatures[['Ductal cells']])
```

### Metanephric cells
```{r}
metanephric_closest_genes <- lapply(c("Metanephric cells"), find_closest)
metanephric_drivers <- intersect(metanephric_closest_genes[[1]], new_signatures[['Metanephric cells']])
```

### Intestinal epithelial cells
```{r}
intestinal_closest_genes <- lapply(c("Intestinal epithelial cells"), find_closest)
intestinal_drivers <- intersect(intestinal_closest_genes[[1]], new_signatures[['Intestinal epithelial cells']])
```

### Ciliated epithelial cells
```{r}
ciliated_closest_genes <- lapply(c("Ciliated epithelial cells"), find_closest)
ciliated_drivers <- intersect(ciliated_closest_genes[[1]], new_signatures[['Ciliated epithelial cells']])
```

### Bronchiolar and alveolar epithelial cells
```{r}
bronchiolar_closest_genes <- lapply(c("Bronchiolar and alveolar epithelial cells"), find_closest)
bronchiolar_drivers <- intersect(bronchiolar_closest_genes[[1]], new_signatures[['Bronchiolar and alveolar epithelial cells']])
```

### Ureteric bud cells
```{r}
ureteric_closest_genes <- lapply(c("Ureteric bud cells"), find_closest)
ureteric_drivers <- intersect(ureteric_closest_genes[[1]], new_signatures[['Ureteric bud cells']])
```

## Other
### Epicardial Fat Cells
```{r}
epicardial_closest_genes <- lapply(c("Epicardial fat cells"), find_closest)
epicardial_drivers <- intersect(epicardial_closest_genes[[1]], new_signatures[['Epicardial fat cells']])
```

### Erythroblasts
```{r}
erythroblast_closest_genes <- lapply(c("Erythroblasts"), find_closest)
erythroblast_drivers <- intersect(erythroblast_closest_genes[[1]], new_signatures[['Erythroblasts']])
```

### Skeletal muscle cells
```{r}
skeletal_closest_genes <- lapply(c("Skeletal muscle cells"), find_closest)
skeletal_drivers <- intersect(skeletal_closest_genes[[1]], new_signatures[['Skeletal muscle cells']])
```

# GSEA
### Housekeeping
Translate the FCA's HGNC names to ENSG
```{r}
gtf_loc <- "/project2/gilad/jpopp/ebQTL/data/gencode/gencode.hg38.filtered.gtf"
pull_gene_type <- function(attr) {
  str_split(attr, "\"")[[1]][11]
}
pull_gene_name <- function(attr) {
  str_split(attr, "\"")[[1]][15]
}
pull_gene_id <- function(attr) {
  str_split(attr, "\"")[[1]][3]
}
gencode <- vroom(gtf_loc, col_names=c("seqname", "source", "feature", "start", "end", "score", "strand", "frame", "attribute"), skip=5) %>%
  filter(seqname %in% paste0("chr", seq(1, 22))) %>%
  filter(feature == "gene") %>%
  mutate(type=map_chr(attribute, pull_gene_type)) %>%
  mutate(hgnc=map_chr(attribute, pull_gene_name)) %>%
  mutate(ensg=map_chr(attribute, pull_gene_id)) %>%
  filter(type=="protein_coding")

## Cell type enrichment analysis
genesets_celltype <- gene_sets_human$gene_set_info %>%
  filter(database == "MSigDB-C8") %>%
  pull(id)
genesets_celltype_info <- gene_sets_human$gene_set_info %>%
  filter(database=="MSigDB-C8")

# Background gene list (in both HGNC and ENSG format)
bg.ensg <- gencode %>%
  filter(hgnc %in% hvg.fca_subsampled) %>%
  pull(ensg) %>%
  intersect(rownames(gene_sets_human$gene_sets)) %>%
  sort
bg.hgnc <- gencode %>%
  filter(ensg %in% bg.ensg) %>%
  arrange(ensg) %>%
  pull(hgnc)
```

Define the main datasets that I plan to use for GSEA
- One for celltype gene sets
- One for pathway gene sets

### Celltype Analysis
```{r}
X_celltype <- gene_sets_human$gene_sets[bg.ensg, genesets_celltype]
X_celltype <- X_celltype[,colSums(X_celltype) >= 10]

logistic_gsea <- function(t) {
  y_t <- bg.hgnc %in% new_signatures[[t]]
  
  logistic.fit <- gseasusie::fit_logistic_susie_veb_boost(X_celltype, y_t, L=10)
  ora <- gseasusie::fit_ora(X_celltype, y_t)
  
  if (!is.null(logistic.fit$sets$cs)) {
    cs <- pivot_longer(as_tibble(logistic.fit$sets$cs), everything(), names_to="component", values_to="index")
  } else {
    cs <- tibble(component=character(), index=numeric())
  }
  
  credible_sets <- as_tibble(logistic.fit$pip, rownames="geneSet") %>%
    dplyr::rename(pip=value) %>%
    rowid_to_column("index") %>%
    left_join(cs, by="index") %>%
    left_join(ora, by="geneSet") %>%
    mutate(fca_type=!!t)
}

gsea_celltypes <- map_dfr(cell.types, logistic_gsea) %>%
  left_join(select(genesets_celltype_info, c(name, id, description_brief)), by=c("geneSet"="id")) %>%
  relocate(name, fca_type, pip, component, pFishersExact)

susie_gsea_celltypes <- gsea_celltypes %>%
  drop_na(component)

#write_tsv(gsea_celltypes, "/project2/gilad/jpopp/ebQTL/data/fca/cellid_genesets_celltype_gsea.tsv")
```

### Pathway Analysis
```{r}
genesets_pathways <- gene_sets_human$gene_set_info %>%
  filter(database == "BioSystems-kegg") %>%
  pull(id)
genesets_pathways_info <- gene_sets_human$gene_set_info %>%
  filter(database=="BioSystems-kegg")

X_pathways <- gene_sets_human$gene_sets[bg.ensg, genesets_pathways]
X_pathways <- X_pathways[,colSums(X_pathways) >= 5]

logistic_gsea_pathways <- function(t) {
  y_t <- bg.hgnc %in% new_signatures[[t]]
  
  logistic.fit <- gseasusie::fit_logistic_susie_veb_boost(X_pathways, y_t, L=10)
  ora <- gseasusie::fit_ora(X_pathways, y_t)
  
  if (!is.null(logistic.fit$sets$cs)) {
    cs <- pivot_longer(as_tibble(logistic.fit$sets$cs), everything(), names_to="component", values_to="index")
  } else {
    cs <- tibble(component=character(), index=numeric())
  }
  
  credible_sets <- as_tibble(logistic.fit$pip, rownames="geneSet") %>%
    dplyr::rename(pip=value) %>%
    rowid_to_column("index") %>%
    left_join(cs, by="index") %>%
    left_join(ora, by="geneSet") %>%
    mutate(fca_type=!!t)
}

gsea_pathways <- map_dfr(cell.types, logistic_gsea_pathways) %>%
  left_join(select(genesets_pathways_info, c(name, id, description_brief)), by=c("geneSet"="id")) %>%
  relocate(name, fca_type, pip, component, pFishersExact)

susie_gsea_pathways <- gsea_pathways %>%
  drop_na(component)

#write_tsv(gsea_pathways, "/project2/gilad/jpopp/ebQTL/data/fca/cellid_genesets_pathways_gsea.tsv")
```



# Troubleshooting with GSEA

## Immune Cells
### Megakaryocytes
#### Pathways
The original megakaryocyte gene signature displayed enrichment for platelet activation
```{r}
gsea_pathways %>%
  filter(fca_type == "Megakaryocytes") %>%
  arrange(desc(pip))
```

This same enrichment appears among the genes which ended up driving the actual annotation
```{r}
y_megakaryocyte <- bg.hgnc %in% megakaryocyte_drivers

logistic.fit.mega.pathway <- gseasusie::fit_logistic_susie_veb_boost(X_pathways, y_megakaryocyte, L=10)
ora.mega.pathway <- gseasusie::fit_ora(X_pathways, y_megakaryocyte)

if (!is.null(logistic.fit.mega.pathway$sets$cs)) {
  cs.mega.pathway <- pivot_longer(as_tibble(logistic.fit.mega.pathway$sets$cs), everything(), names_to="component", values_to="index")
} else {
  cs.mega.pathway <- tibble(component=character(), index=numeric())
}

credible_sets_mega_pathway <- as_tibble(logistic.fit.mega.pathway$pip, rownames="geneSet") %>%
  dplyr::rename(pip=value) %>%
  rowid_to_column("index") %>%
  left_join(cs.mega.pathway, by="index") %>%
  left_join(ora.mega.pathway, by="geneSet")

gsea_pathways_megakaryocyte <- credible_sets_mega_pathway %>%
  left_join(select(genesets_pathways_info, c(name, id, description_brief)), by=c("geneSet"="id")) %>%
  relocate(name, pip, component, pFishersExact)

gsea_pathways_megakaryocyte %>%
  arrange(desc(pip))
```

#### Celltypes
The original megakaryocyte gene signature also displayed enrichment for a few RBC-related cell types
```{r}
gsea_celltypes %>%
  filter(fca_type == "Megakaryocytes") %>%
  arrange(desc(pip))
```

Two of these pop up again among the genes driving the annotation
```{r}
logistic.fit.mega.celltype <- gseasusie::fit_logistic_susie_veb_boost(X_celltype, y_megakaryocyte, L=10)
ora.mega.celltype <- gseasusie::fit_ora(X_celltype, y_megakaryocyte)

if (!is.null(logistic.fit.mega.celltype$sets$cs)) {
  cs.mega.celltype <- pivot_longer(as_tibble(logistic.fit.mega.celltype$sets$cs), everything(), names_to="component", values_to="index")
} else {
  cs.mega.celltype <- tibble(component=character(), index=numeric())
}

credible_sets_mega_celltype <- as_tibble(logistic.fit.mega.celltype$pip, rownames="geneSet") %>%
  dplyr::rename(pip=value) %>%
  rowid_to_column("index") %>%
  left_join(cs.mega.celltype, by="index") %>%
  left_join(ora.mega.celltype, by="geneSet")

gsea_celltypes_megakaryocyte <- credible_sets_mega_celltype %>%
  left_join(select(genesets_celltype_info, c(name, id, description_brief)), by=c("geneSet"="id")) %>%
  relocate(name, pip, component, pFishersExact)

gsea_celltypes_megakaryocyte %>%
  arrange(desc(pip))
```

Does this hold for the genes that aren't just in the megakaryocyte signature, but like all EB megakaryocyte marker genes?
```{r}
y_megakaryocyte_full <- bg.hgnc %in% megakaryocyte_closest_genes[[1]]
logistic.fit.mega.full.celltype <- gseasusie::fit_logistic_susie_veb_boost(X_celltype, y_megakaryocyte_full, L=10)
ora.mega.full.celltype <- gseasusie::fit_ora(X_celltype, y_megakaryocyte_full)

if (!is.null(logistic.fit.mega.full.celltype$sets$cs)) {
  cs.mega.full.celltype <- pivot_longer(as_tibble(logistic.fit.mega.full.celltype$sets$cs), everything(), names_to="component", values_to="index")
} else {
  cs.mega.full.celltype <- tibble(component=character(), index=numeric())
}

credible_sets_mega_full_celltype <- as_tibble(logistic.fit.mega.full.celltype$pip, rownames="geneSet") %>%
  dplyr::rename(pip=value) %>%
  rowid_to_column("index") %>%
  left_join(cs.mega.full.celltype, by="index") %>%
  left_join(ora.mega.full.celltype, by="geneSet")

gsea_celltypes_megakaryocyte_full <- credible_sets_mega_full_celltype %>%
  left_join(select(genesets_celltype_info, c(name, id, description_brief)), by=c("geneSet"="id")) %>%
  relocate(name, pip, component, pFishersExact)

gsea_celltypes_megakaryocyte_full %>%
  arrange(desc(pip))
```
It's pretty important that the retinal microglia are in there, but this looks assuring


### Lymphoid cells
#### Pathways
The original lymphoid gene signature displayed enrichment for T cell receptor signaling
```{r}
gsea_pathways %>%
  filter(fca_type == "Lymphoid cells") %>%
  arrange(desc(pip))
```

This same enrichment *does not appear* among the genes which ended up driving the actual annotation
```{r}
y_lymphoid <- bg.hgnc %in% lymphoid_drivers

logistic.fit.lymph.pathway <- gseasusie::fit_logistic_susie_veb_boost(X_pathways, y_lymphoid, L=10)
ora.lymph.pathway <- gseasusie::fit_ora(X_pathways, y_lymphoid)

if (!is.null(logistic.fit.lymph.pathway$sets$cs)) {
  cs.lymph.pathway <- pivot_longer(as_tibble(logistic.fit.lymph.pathway$sets$cs), everything(), names_to="component", values_to="index")
} else {
  cs.lymph.pathway <- tibble(component=character(), index=numeric())
}

credible_sets_lymph_pathway <- as_tibble(logistic.fit.lymph.pathway$pip, rownames="geneSet") %>%
  dplyr::rename(pip=value) %>%
  rowid_to_column("index") %>%
  left_join(cs.lymph.pathway, by="index") %>%
  left_join(ora.lymph.pathway, by="geneSet")

gsea_pathways_lymphoid <- credible_sets_lymph_pathway %>%
  left_join(select(genesets_pathways_info, c(name, id, description_brief)), by=c("geneSet"="id")) %>%
  relocate(name, pip, component, pFishersExact)

gsea_pathways_lymphoid %>%
  filter(name=="T cell receptor signaling pathway") %>%
  relocate(name, pip, pFishersExact, geneListSize, geneSetSize,)
```

#### Celltypes
The original lymphoid cell gene signature also displayed enrichment for a few WBC-related cell types,
as well as a large intestine mesenchymal cell type
```{r}
gsea_celltypes %>%
  filter(fca_type == "Lymphoid cells") %>%
  arrange(component)
```

Do the genes in that intestinal cell type overlap with those of the immune cell types?
```{r}
gao_genes <- bg.hgnc[X_celltype[,filter(genesets_celltype_info, name=="GAO_LARGE_INTESTINE_ADULT_CI_MESENCHYMAL_CELLS")$id[1]]==1]
cui_genes <- bg.hgnc[X_celltype[,filter(genesets_celltype_info, name=="CUI_DEVELOPING_HEART_C9_B_T_CELL")$id[1]]==1]

print(gao_genes)
print(cui_genes)
print(intersect(gao_genes, cui_genes))
```

The cell type analysis does pop up again
```{r}
logistic.fit.lymph.celltype <- gseasusie::fit_logistic_susie_veb_boost(X_celltype, y_lymphoid, L=10)
ora.lymph.celltype <- gseasusie::fit_ora(X_celltype, y_lymphoid)

if (!is.null(logistic.fit.lymph.celltype$sets$cs)) {
  cs.lymph.celltype <- pivot_longer(as_tibble(logistic.fit.lymph.celltype$sets$cs), everything(), names_to="component", values_to="index")
} else {
  cs.lymph.celltype <- tibble(component=character(), index=numeric())
}

credible_sets_lymph_celltype <- as_tibble(logistic.fit.lymph.celltype$pip, rownames="geneSet") %>%
  dplyr::rename(pip=value) %>%
  rowid_to_column("index") %>%
  left_join(cs.lymph.celltype, by="index") %>%
  left_join(ora.lymph.celltype, by="geneSet")

gsea_celltypes_lymphoid <- credible_sets_lymph_celltype %>%
  left_join(select(genesets_celltype_info, c(name, id, description_brief)), by=c("geneSet"="id")) %>%
  relocate(name, pip, component, pFishersExact)

gsea_celltypes_lymphoid %>%
  arrange(desc(pip)) %>%
  relocate(name, pip, component, pFishersExact, geneListSize, geneSetSize, overlap)
```

What if we don't just look at the genes driving the annotation, but all marker genes for this cluster?
```{r}
y_lymphoid_full <- bg.hgnc %in% lymphoid_closest_genes[[1]]

logistic.fit.lymph.full.celltype <- gseasusie::fit_logistic_susie_veb_boost(X_celltype, y_lymphoid_full, L=10)
ora.lymph.full.celltype <- gseasusie::fit_ora(X_celltype, y_lymphoid_full)

if (!is.null(logistic.fit.lymph.full.celltype$sets$cs)) {
  cs.lymph.full.celltype <- pivot_longer(as_tibble(logistic.fit.lymph.full.celltype$sets$cs), everything(), names_to="component", values_to="index")
} else {
  cs.lymph.full.celltype <- tibble(component=character(), index=numeric())
}

credible_sets_lymph_full_celltype <- as_tibble(logistic.fit.lymph.full.celltype$pip, rownames="geneSet") %>%
  dplyr::rename(pip=value) %>%
  rowid_to_column("index") %>%
  left_join(cs.lymph.full.celltype, by="index") %>%
  left_join(ora.lymph.full.celltype, by="geneSet")

gsea_celltypes_lymphoid_full <- credible_sets_lymph_full_celltype %>%
  left_join(select(genesets_celltype_info, c(name, id, description_brief)), by=c("geneSet"="id")) %>%
  relocate(name, pip, component, pFishersExact)

gsea_celltypes_lymphoid_full %>%
  arrange(desc(pip)) %>%
  relocate(name, pip, component, pFishersExact, geneListSize, geneSetSize, overlap)
```

### Erythroblasts
```{r}
gsea_celltypes %>%
  filter(fca_type == "Erythroblasts") %>%
  arrange(desc(pip))
```
First, not just for the drivers, but for the full signature of the EB erythroblasts specifically, what cell types are we looking at?
```{r}
y_erythroblast <- bg.hgnc %in% erythroblast_closest_genes[[1]]

logistic.fit.erythroblast.celltype <- gseasusie::fit_logistic_susie_veb_boost(X_celltype, y_erythroblast, L=10)
ora.erythroblast.celltype <- gseasusie::fit_ora(X_celltype, y_erythroblast)

if (!is.null(logistic.fit.erythroblast.celltype$sets$cs)) {
  cs.erythroblast.celltype <- pivot_longer(as_tibble(logistic.fit.erythroblast.celltype$sets$cs), everything(), names_to="component", values_to="index")
} else {
  cs.erythroblast.celltype <- tibble(component=character(), index=numeric())
}

credible_sets_erythroblast_celltype <- as_tibble(logistic.fit.erythroblast.celltype$pip, rownames="geneSet") %>%
  dplyr::rename(pip=value) %>%
  rowid_to_column("index") %>%
  left_join(cs.erythroblast.celltype, by="index") %>%
  left_join(ora.erythroblast.celltype, by="geneSet")

gsea_celltypes_erythroblast <- credible_sets_erythroblast_celltype %>%
  left_join(select(genesets_celltype_info, c(name, id, description_brief)), by=c("geneSet"="id")) %>%
  relocate(name, pip, component, pFishersExact)

gsea_celltypes_erythroblast %>%
  arrange(desc(pip))
```
1. Which genes from our erythroblast signature have previously been confirmed to be erythroblast markers?
```{r}
hay_erythroblast <- tibble(ensg=names(X_celltype[,"M39197"])[X_celltype[,"M39197"]==1]) %>%
  left_join(gencode, by="ensg") %>%
  pull(hgnc)

erythroblast_overlap <- intersect(hay_erythroblast, new_signatures[['Erythroblasts']])

erythroblast_overlap
```
Yes, plenty of overlap between ours and an existing erythroblast signature.

Do those same genes appear in the Manno progenitor basal gene set?
```{r}
manno_hprogbp <- tibble(ensg=names(X_celltype[,"M39059"])[X_celltype[,"M39059"]==1]) %>%
  left_join(gencode, by="ensg") %>%
  pull(hgnc)

bp_overlap <- intersect(manno_hprogbp, new_signatures[['Erythroblasts']])

intersect(erythroblast_overlap, bp_overlap)
```
Just the one, which isn't overly tissue-specific I don't think

Now let's look at (1) the genes driving the annotation of these cells, and *their* overlap with BP
```{r}
bp_problem_genes <- intersect(erythroblast_closest_genes[[1]], manno_hprogbp)
bp_problem_genes
```
ASPM, MKI67, TOP2A, CENPE, CENPF, these are lots of genes involved in mitosis and 
maintaining chromosomal stability. 

NELL2 is the only one that's really brain specific here, and 1 in 14 isn't really cause
for concern

Finally, I'll check whether there's overlap with real erythroblast genes on the actual EB side
```{r}
intersect(erythroblast_overlap, erythroblast_closest_genes[[1]])
```
Seems totally reasonable, FECH is very specifically expressed in blood and is involved in heme synthesis, 
and CPOX is involved in heme production too. 



## Epithelial Cells
### Acinar cells
```{r}
gsea_celltypes %>%
  filter(fca_type == "Acinar cells") %>%
  arrange(desc(pip))
```

First, not just for the drivers, but for the full signature of the EB acinar cells specifically, what cell types are we looking at?
```{r}
y_acinar <- bg.hgnc %in% acinar_closest_genes[[1]]

logistic.fit.acinar.celltype <- gseasusie::fit_logistic_susie_veb_boost(X_celltype, y_acinar, L=10)
ora.acinar.celltype <- gseasusie::fit_ora(X_celltype, y_acinar)

if (!is.null(logistic.fit.acinar.celltype$sets$cs)) {
  cs.acinar.celltype <- pivot_longer(as_tibble(logistic.fit.acinar.celltype$sets$cs), everything(), names_to="component", values_to="index")
} else {
  cs.acinar.celltype <- tibble(component=character(), index=numeric())
}

credible_sets_acinar_celltype <- as_tibble(logistic.fit.acinar.celltype$pip, rownames="geneSet") %>%
  dplyr::rename(pip=value) %>%
  rowid_to_column("index") %>%
  left_join(cs.acinar.celltype, by="index") %>%
  left_join(ora.acinar.celltype, by="geneSet")

gsea_celltypes_acinar <- credible_sets_acinar_celltype %>%
  left_join(select(genesets_celltype_info, c(name, id, description_brief)), by=c("geneSet"="id")) %>%
  relocate(name, pip, component, pFishersExact)

gsea_celltypes_acinar %>%
  arrange(desc(pip))
```
**These are neuronal cell types**

Does our signature look ok?
```{r}
muraro_acinar <- tibble(ensg=names(X_celltype[,"M39174"])[X_celltype[,"M39174"]==1]) %>%
  left_join(gencode, by="ensg") %>%
  pull(hgnc)

acinar_overlap <- intersect(muraro_acinar, new_signatures[['Acinar cells']])

acinar_overlap
```

Yeah these look fine, CPA1 is very pancreas-specific

It's not these acinar-confirmed genes that are driving the mistaken annotation as CTX for the acinar centroid in EBs
```{r}
fan_ctx <- tibble(ensg=names(X_celltype[,"M39036"])[X_celltype[,"M39036"]==1]) %>%
  left_join(gencode, by="ensg") %>%
  pull(hgnc)

ctx_overlap <- intersect(fan_ctx, new_signatures[['Acinar cells']])

intersect(acinar_overlap, ctx_overlap)
```

```{r}
ctx_problem_genes <- intersect(acinar_closest_genes[[1]], fan_ctx)
ctx_problem_genes
```
The genes that are driving that cortical annotation are bad cortical markers

CENPE, CENPF, TOP2A - not good cortical markers

Is there intersection with the good acinar gene set?
```{r}
intersect(acinar_closest_genes[[1]], muraro_acinar)
```
CPA2 is the only one out of these that GTEx supports as a pancreatic marker

### Ductal cells
```{r}
gsea_celltypes %>%
  filter(fca_type == "Ductal cells") %>%
  arrange(desc(pip))
```

```{r}
y_ductal <- bg.hgnc %in% ductal_closest_genes[[1]]

logistic.fit.ductal.celltype <- gseasusie::fit_logistic_susie_veb_boost(X_celltype, y_ductal, L=10)
ora.ductal.celltype <- gseasusie::fit_ora(X_celltype, y_ductal)

if (!is.null(logistic.fit.ductal.celltype$sets$cs)) {
  cs.ductal.celltype <- pivot_longer(as_tibble(logistic.fit.ductal.celltype$sets$cs), everything(), names_to="component", values_to="index")
} else {
  cs.ductal.celltype <- tibble(component=character(), index=numeric())
}

credible_sets_ductal_celltype <- as_tibble(logistic.fit.ductal.celltype$pip, rownames="geneSet") %>%
  dplyr::rename(pip=value) %>%
  rowid_to_column("index") %>%
  left_join(cs.ductal.celltype, by="index") %>%
  left_join(ora.ductal.celltype, by="geneSet")

gsea_celltypes_ductal <- credible_sets_ductal_celltype %>%
  left_join(select(genesets_celltype_info, c(name, id, description_brief)), by=c("geneSet"="id")) %>%
  relocate(name, pip, component, pFishersExact)

gsea_celltypes_ductal %>%
  arrange(desc(pip))
```

### Metanephric cells
```{r}
gsea_celltypes %>%
  filter(fca_type == "Metanephric cells") %>%
  arrange(desc(pip))
```

```{r}
y_metanephric <- bg.hgnc %in% metanephric_closest_genes[[1]]

logistic.fit.metanephric.celltype <- gseasusie::fit_logistic_susie_veb_boost(X_celltype, y_metanephric, L=10)
ora.metanephric.celltype <- gseasusie::fit_ora(X_celltype, y_metanephric)

if (!is.null(logistic.fit.metanephric.celltype$sets$cs)) {
  cs.metanephric.celltype <- pivot_longer(as_tibble(logistic.fit.metanephric.celltype$sets$cs), everything(), names_to="component", values_to="index")
} else {
  cs.metanephric.celltype <- tibble(component=character(), index=numeric())
}

credible_sets_metanephric_celltype <- as_tibble(logistic.fit.metanephric.celltype$pip, rownames="geneSet") %>%
  dplyr::rename(pip=value) %>%
  rowid_to_column("index") %>%
  left_join(cs.metanephric.celltype, by="index") %>%
  left_join(ora.metanephric.celltype, by="geneSet")

gsea_celltypes_metanephric <- credible_sets_metanephric_celltype %>%
  left_join(select(genesets_celltype_info, c(name, id, description_brief)), by=c("geneSet"="id")) %>%
  relocate(name, pip, component, pFishersExact)

gsea_celltypes_metanephric %>%
  arrange(desc(pip))
```

### Intestinal epithelial cells
```{r}
gsea_celltypes %>%
  filter(fca_type == "Intestinal epithelial cells") %>%
  arrange(desc(pip))
```

```{r}
y_intestinal <- bg.hgnc %in% intestinal_closest_genes[[1]]

logistic.fit.intestinal.celltype <- gseasusie::fit_logistic_susie_veb_boost(X_celltype, y_intestinal, L=10)
ora.intestinal.celltype <- gseasusie::fit_ora(X_celltype, y_intestinal)

if (!is.null(logistic.fit.intestinal.celltype$sets$cs)) {
  cs.intestinal.celltype <- pivot_longer(as_tibble(logistic.fit.intestinal.celltype$sets$cs), everything(), names_to="component", values_to="index")
} else {
  cs.intestinal.celltype <- tibble(component=character(), index=numeric())
}

credible_sets_intestinal_celltype <- as_tibble(logistic.fit.intestinal.celltype$pip, rownames="geneSet") %>%
  dplyr::rename(pip=value) %>%
  rowid_to_column("index") %>%
  left_join(cs.intestinal.celltype, by="index") %>%
  left_join(ora.intestinal.celltype, by="geneSet")

gsea_celltypes_intestinal <- credible_sets_intestinal_celltype %>%
  left_join(select(genesets_celltype_info, c(name, id, description_brief)), by=c("geneSet"="id")) %>%
  relocate(name, pip, component, pFishersExact)

gsea_celltypes_intestinal %>%
  arrange(desc(pip))
```

### Ciliated epithelial cells
```{r}
gsea_celltypes %>%
  filter(fca_type == "Ciliated epithelial cells") %>%
  arrange(desc(pip))
```

```{r}
y_ciliated <- bg.hgnc %in% ciliated_closest_genes[[1]]

logistic.fit.ciliated.celltype <- gseasusie::fit_logistic_susie_veb_boost(X_celltype, y_ciliated, L=10)
ora.ciliated.celltype <- gseasusie::fit_ora(X_celltype, y_ciliated)

if (!is.null(logistic.fit.ciliated.celltype$sets$cs)) {
  cs.ciliated.celltype <- pivot_longer(as_tibble(logistic.fit.ciliated.celltype$sets$cs), everything(), names_to="component", values_to="index")
} else {
  cs.ciliated.celltype <- tibble(component=character(), index=numeric())
}

credible_sets_ciliated_celltype <- as_tibble(logistic.fit.ciliated.celltype$pip, rownames="geneSet") %>%
  dplyr::rename(pip=value) %>%
  rowid_to_column("index") %>%
  left_join(cs.ciliated.celltype, by="index") %>%
  left_join(ora.ciliated.celltype, by="geneSet")

gsea_celltypes_ciliated <- credible_sets_ciliated_celltype %>%
  left_join(select(genesets_celltype_info, c(name, id, description_brief)), by=c("geneSet"="id")) %>%
  relocate(name, pip, component, pFishersExact)

gsea_celltypes_ciliated %>%
  arrange(desc(pip))
```

### Bronchiolar and alveolar epithelial cells
```{r}
gsea_celltypes %>%
  filter(fca_type == "Bronchiolar and alveolar epithelial cells") %>%
  arrange(desc(pip))
```

```{r}
y_bronchiolar <- bg.hgnc %in% bronchiolar_closest_genes[[1]]

logistic.fit.bronchiolar.celltype <- gseasusie::fit_logistic_susie_veb_boost(X_celltype, y_bronchiolar, L=10)
ora.bronchiolar.celltype <- gseasusie::fit_ora(X_celltype, y_bronchiolar)

if (!is.null(logistic.fit.bronchiolar.celltype$sets$cs)) {
  cs.bronchiolar.celltype <- pivot_longer(as_tibble(logistic.fit.bronchiolar.celltype$sets$cs), everything(), names_to="component", values_to="index")
} else {
  cs.bronchiolar.celltype <- tibble(component=character(), index=numeric())
}

credible_sets_bronchiolar_celltype <- as_tibble(logistic.fit.bronchiolar.celltype$pip, rownames="geneSet") %>%
  dplyr::rename(pip=value) %>%
  rowid_to_column("index") %>%
  left_join(cs.bronchiolar.celltype, by="index") %>%
  left_join(ora.bronchiolar.celltype, by="geneSet")

gsea_celltypes_bronchiolar <- credible_sets_bronchiolar_celltype %>%
  left_join(select(genesets_celltype_info, c(name, id, description_brief)), by=c("geneSet"="id")) %>%
  relocate(name, pip, component, pFishersExact)

gsea_celltypes_bronchiolar %>%
  arrange(desc(pip))
```

### Ureteric bud cells
```{r}
gsea_celltypes %>%
  filter(fca_type == "Ureteric bud cells") %>%
  arrange(desc(pip))
```

```{r}
y_ureteric <- bg.hgnc %in% ureteric_closest_genes[[1]]

logistic.fit.ureteric.celltype <- gseasusie::fit_logistic_susie_veb_boost(X_celltype, y_ureteric, L=10)
ora.ureteric.celltype <- gseasusie::fit_ora(X_celltype, y_ureteric)

if (!is.null(logistic.fit.ureteric.celltype$sets$cs)) {
  cs.ureteric.celltype <- pivot_longer(as_tibble(logistic.fit.ureteric.celltype$sets$cs), everything(), names_to="component", values_to="index")
} else {
  cs.ureteric.celltype <- tibble(component=character(), index=numeric())
}

credible_sets_ureteric_celltype <- as_tibble(logistic.fit.ureteric.celltype$pip, rownames="geneSet") %>%
  dplyr::rename(pip=value) %>%
  rowid_to_column("index") %>%
  left_join(cs.ureteric.celltype, by="index") %>%
  left_join(ora.ureteric.celltype, by="geneSet")

gsea_celltypes_ureteric <- credible_sets_ureteric_celltype %>%
  left_join(select(genesets_celltype_info, c(name, id, description_brief)), by=c("geneSet"="id")) %>%
  relocate(name, pip, component, pFishersExact)

gsea_celltypes_ureteric %>%
  arrange(pFishersExact)
```



## Other
### Epicardial fat cells
```{r}
gsea_celltypes %>%
  filter(fca_type == "Epicardial fat cells") %>%
  arrange(desc(pip))
```

First, not just for the drivers, but for the full signature of the EB epicardial fat cells specifically, what cell types are we looking at?
```{r}
y_epicardial <- bg.hgnc %in% epicardial_closest_genes[[1]]

logistic.fit.epicardial.celltype <- gseasusie::fit_logistic_susie_veb_boost(X_celltype, y_epicardial, L=10)
ora.epicardial.celltype <- gseasusie::fit_ora(X_celltype, y_epicardial)

if (!is.null(logistic.fit.epicardial.celltype$sets$cs)) {
  cs.epicardial.celltype <- pivot_longer(as_tibble(logistic.fit.epicardial.celltype$sets$cs), everything(), names_to="component", values_to="index")
} else {
  cs.epicardial.celltype <- tibble(component=character(), index=numeric())
}

credible_sets_epicardial_celltype <- as_tibble(logistic.fit.epicardial.celltype$pip, rownames="geneSet") %>%
  dplyr::rename(pip=value) %>%
  rowid_to_column("index") %>%
  left_join(cs.epicardial.celltype, by="index") %>%
  left_join(ora.epicardial.celltype, by="geneSet")

gsea_celltypes_epicardial <- credible_sets_epicardial_celltype %>%
  left_join(select(genesets_celltype_info, c(name, id, description_brief)), by=c("geneSet"="id")) %>%
  relocate(name, pip, component, pFishersExact)

gsea_celltypes_epicardial %>%
  arrange(desc(pip))
```

UGH. Is our epicardium signature that we learned any good?
```{r}
cui_epicardium <- tibble(ensg=names(X_celltype[,"M39303"])[X_celltype[,"M39303"]==1]) %>%
  left_join(gencode, by="ensg") %>%
  pull(hgnc)

epicardium_overlap <- intersect(cui_epicardium, new_signatures[['Epicardial fat cells']])

epicardium_overlap
```
Yes, plenty of overlap between this and an existing epicardium signature.

Do those same genes appear in the RPE gene set?
```{r}
hu_rpe <- tibble(ensg=names(X_celltype[,"M39271"])[X_celltype[,"M39271"]==1]) %>%
  left_join(gencode, by="ensg") %>%
  pull(hgnc)

rpe_overlap <- intersect(hu_rpe, new_signatures[['Epicardial fat cells']])

intersect(epicardium_overlap, rpe_overlap)
```
Yes! Some do, and it looks like that's fine

SEMA3C - important gene involved in cardiovascular development, as well as axon growth and guidance.
CCDC80 - cell adhesion & matrix assembly
EZR - ezrin, also epithelial cells

Now let's look at (1) the genes driving the annotation of these cells, and *their* overlap with RPE
```{r}
rpe_problem_genes <- intersect(epicardial_closest_genes[[1]], hu_rpe)
rpe_problem_genes
```

These actually look more like epicardial markers than a retinal cell type, I don't think 
there's a huge problem with the annotation.

MYL9, - specifically in artery tissues
KRT8 - broadly in epithelial cells

And we do see overlap with epicardial cells right?
```{r}
intersect(epicardial_closest_genes[[1]], cui_epicardium)
```

Are there just two distinct groups of epicardial fat cells?
```{r fig.width=8, fig.height=4}
emb_viz <- select(cell_emb, c(MCA_1, MCA_2, celltype)) %>%
  drop_na() %>%
  group_by(celltype) %>%
  sample_n(min(n(), 200))

emb_viz %>%
  ggplot(aes(x=MCA_1, y=MCA_2, color=celltype)) +
  geom_point(size=0.1)

emb_viz %>%
  mutate(celltype_epi = if_else(celltype=="Epicardial fat cells", "Epicardial fat cells", "Other")) %>%
  mutate(celltype_epi=factor(celltype_epi, levels=c("Other", "Epicardial fat cells"))) %>%
  ggplot(aes(x=MCA_1, y=MCA_2)) +
  geom_point(aes(color=celltype_epi), size=0.1)  +
  scale_color_manual(values=c("light grey", hue_pal()(1)))
```
```{r}
fat_cells <- eb$celltype == "Epicardial fat cells"
fat_cells[is.na(fat_cells)] <- FALSE
fat_cells_embedded <- runPCA(eb[,fat_cells])

fat_cells_marked <- as_tibble(reducedDim(fat_cells_embedded, "PCA"), rownames="cell") %>%
  mutate(APOE=logcounts(fat_cells_embedded)["APOE",]) %>%
  mutate(PAX6=logcounts(fat_cells_embedded)["PAX6",]) %>%
  mutate(MYL9=logcounts(fat_cells_embedded)["MYL9",])
  

fat_cells_marked %>%
  ggplot(aes(x=PC1, y=PC2, color=APOE)) +
  geom_point(size=0.5)

fat_cells_marked %>%
  ggplot(aes(x=PC1, y=PC2, color=PAX6)) +
  geom_point(size=0.5)

fat_cells_marked %>%
  ggplot(aes(x=APOE, y=PAX6)) +
  geom_point(size=0.5)

fat_cells_marked %>%
  ggplot(aes(x=MYL9, y=PAX6)) +
  geom_point(size=0.5)
```

```{r fig.width=3, fig.height=3}
#fat_cells_embedded <- runUMAP(fat_cells_embedded)

plot_density(fat_cells_embedded, c("APOE", "MYL9", "PAX6", "SULF1"))
```
```{r}
fat_cells_marked %>%
  group_by(group) %>%
  dplyr::count()
```


### Skeletal muscle cells
```{r}
gsea_celltypes %>%
  filter(fca_type == "Skeletal muscle cells") %>%
  arrange(desc(overlap))
```
Not sure whether there's anything better in the cell types, cardiomyocyte is a muscle cell type so this is ok for now

Next, not just for the drivers, but for the full signature of the EB skeletal muscle cells specifically, what cell types are we looking at?
```{r}
y_skeletal <- bg.hgnc %in% skeletal_closest_genes[[1]]

logistic.fit.skeletal.celltype <- gseasusie::fit_logistic_susie_veb_boost(X_celltype, y_skeletal, L=10)
ora.skeletal.celltype <- gseasusie::fit_ora(X_celltype, y_skeletal)

if (!is.null(logistic.fit.skeletal.celltype$sets$cs)) {
  cs.skeletal.celltype <- pivot_longer(as_tibble(logistic.fit.skeletal.celltype$sets$cs), everything(), names_to="component", values_to="index")
} else {
  cs.skeletal.celltype <- tibble(component=character(), index=numeric())
}

credible_sets_skeletal_celltype <- as_tibble(logistic.fit.skeletal.celltype$pip, rownames="geneSet") %>%
  dplyr::rename(pip=value) %>%
  rowid_to_column("index") %>%
  left_join(cs.skeletal.celltype, by="index") %>%
  left_join(ora.skeletal.celltype, by="geneSet")

gsea_celltypes_skeletal <- credible_sets_skeletal_celltype %>%
  left_join(select(genesets_celltype_info, c(name, id, description_brief)), by=c("geneSet"="id")) %>%
  relocate(name, pip, component, pFishersExact)

gsea_celltypes_skeletal %>%
  arrange(desc(pip))
```
Inconclusive! There's a whole study on skeletal muscle cells that's not reflected here, though.

I'll peek at Cui's smooth muscle for comparison

Is our skeletal muscle signature that we learned any good?
```{r}
cui_muscle <- tibble(ensg=names(X_celltype[,"M39318"])[X_celltype[,"M39318"]==1]) %>%
  left_join(gencode, by="ensg") %>%
  pull(hgnc)

muscle_overlap <- intersect(cui_muscle, new_signatures[['Skeletal muscle cells']])

muscle_overlap
```
None w smooth muscle, how about cardiomyocytes though.
```{r}
cui_myocyte <- tibble(ensg=names(X_celltype[,"M39299"])[X_celltype[,"M39299"]==1]) %>%
  left_join(gencode, by="ensg") %>%
  pull(hgnc)

myocyte_overlap <- intersect(cui_myocyte, new_signatures[['Skeletal muscle cells']])

myocyte_overlap
```
These all look reasonable

Do those same genes appear in the medial neuroblast gene set?
```{r}
manno_hnbm <- tibble(ensg=names(X_celltype[,"M39063"])[X_celltype[,"M39063"]==1]) %>%
  left_join(gencode, by="ensg") %>%
  pull(hgnc)

nbm_overlap <- intersect(manno_hnbm, new_signatures[['Skeletal muscle cells']])

intersect(myocyte_overlap, nbm_overlap)
```
No, different subgroups of the skeletal muscle signature

Now let's look at (1) the genes driving the annotation of these cells, and *their* overlap with NBM
```{r}
nbm_problem_genes <- intersect(skeletal_closest_genes[[1]], manno_hnbm)
nbm_problem_genes
```
Bit of a mixed bag. NEUROD1, NEUROD4, STMN2 seem pretty brain-specific

SRL is definitely skeletal muscle though

FNDC5 is expressed in both brain and skeletal muscle according to GTEx

And then most of these are kind of out of nowhere.

Do we also see real overlap w the myocyte markers?
```{r}
intersect(skeletal_closest_genes[[1]], cui_myocyte)
```
These are reasonable cardiac cell type markers, but there's not many of them.


# Next