{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7d9ceaae-e8d9-4f87-a75f-d3d29dd58bdf",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jpopp/.conda/envs/tensorqtl/lib/python3.11/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: 'rfunc' cannot be imported. R and the 'rpy2' Python package are needed.\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import torch\n",
    "import tensorqtl\n",
    "import numpy as np\n",
    "from tensorqtl import genotypeio, cis, trans\n",
    "from sklearn import preprocessing\n",
    "from sklearn.preprocessing import quantile_transform\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "cebfef9d-cfb2-4a62-b22f-7187c07dc82f",
   "metadata": {},
   "outputs": [],
   "source": [
    "genotype_df = pd.read_csv(\"../../data/trans_qtl_calling/gtex/genotypes_filtered/Vagina.GOBP_TISSUE_DEVELOPMENT.tsv.gz\", \n",
    "                          sep=\"\\t\", index_col='snp')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "9474fd37-d69b-4fb0-8e81-27904e4eeb57",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GTEX-113JC</th>\n",
       "      <th>GTEX-11DXX</th>\n",
       "      <th>GTEX-11EM3</th>\n",
       "      <th>GTEX-11EMC</th>\n",
       "      <th>GTEX-11GSP</th>\n",
       "      <th>GTEX-11I78</th>\n",
       "      <th>GTEX-11P81</th>\n",
       "      <th>GTEX-11TTK</th>\n",
       "      <th>GTEX-11VI4</th>\n",
       "      <th>GTEX-11XUK</th>\n",
       "      <th>...</th>\n",
       "      <th>GTEX-ZP4G</th>\n",
       "      <th>GTEX-ZQG8</th>\n",
       "      <th>GTEX-ZTPG</th>\n",
       "      <th>GTEX-ZV6S</th>\n",
       "      <th>GTEX-ZVT2</th>\n",
       "      <th>GTEX-ZVT3</th>\n",
       "      <th>GTEX-ZXES</th>\n",
       "      <th>GTEX-ZYVF</th>\n",
       "      <th>GTEX-ZYY3</th>\n",
       "      <th>GTEX-ZZPU</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>snp</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1_9919625_C_T_b38</th>\n",
       "      <td>0.531717</td>\n",
       "      <td>0.531717</td>\n",
       "      <td>0.531717</td>\n",
       "      <td>0.531717</td>\n",
       "      <td>0.531717</td>\n",
       "      <td>0.531717</td>\n",
       "      <td>0.531717</td>\n",
       "      <td>-1.536070</td>\n",
       "      <td>0.531717</td>\n",
       "      <td>-1.536070</td>\n",
       "      <td>...</td>\n",
       "      <td>0.531717</td>\n",
       "      <td>0.531717</td>\n",
       "      <td>-1.536070</td>\n",
       "      <td>0.531717</td>\n",
       "      <td>0.531717</td>\n",
       "      <td>0.531717</td>\n",
       "      <td>0.531717</td>\n",
       "      <td>-1.536070</td>\n",
       "      <td>0.531717</td>\n",
       "      <td>0.531717</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1_9932244_A_C_b38</th>\n",
       "      <td>0.520278</td>\n",
       "      <td>0.520278</td>\n",
       "      <td>0.520278</td>\n",
       "      <td>0.520278</td>\n",
       "      <td>0.520278</td>\n",
       "      <td>0.520278</td>\n",
       "      <td>0.520278</td>\n",
       "      <td>-1.347385</td>\n",
       "      <td>0.520278</td>\n",
       "      <td>-1.347385</td>\n",
       "      <td>...</td>\n",
       "      <td>0.520278</td>\n",
       "      <td>0.520278</td>\n",
       "      <td>-1.347385</td>\n",
       "      <td>0.520278</td>\n",
       "      <td>0.520278</td>\n",
       "      <td>0.520278</td>\n",
       "      <td>0.520278</td>\n",
       "      <td>-1.347385</td>\n",
       "      <td>0.520278</td>\n",
       "      <td>0.520278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1_9934281_A_G_b38</th>\n",
       "      <td>0.531717</td>\n",
       "      <td>0.531717</td>\n",
       "      <td>0.531717</td>\n",
       "      <td>0.531717</td>\n",
       "      <td>0.531717</td>\n",
       "      <td>0.531717</td>\n",
       "      <td>0.531717</td>\n",
       "      <td>-1.536070</td>\n",
       "      <td>0.531717</td>\n",
       "      <td>-1.536070</td>\n",
       "      <td>...</td>\n",
       "      <td>0.531717</td>\n",
       "      <td>0.531717</td>\n",
       "      <td>-1.536070</td>\n",
       "      <td>0.531717</td>\n",
       "      <td>0.531717</td>\n",
       "      <td>0.531717</td>\n",
       "      <td>0.531717</td>\n",
       "      <td>-1.536070</td>\n",
       "      <td>0.531717</td>\n",
       "      <td>0.531717</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1_9940342_T_A_b38</th>\n",
       "      <td>0.509803</td>\n",
       "      <td>0.509803</td>\n",
       "      <td>0.509803</td>\n",
       "      <td>0.509803</td>\n",
       "      <td>0.509803</td>\n",
       "      <td>0.509803</td>\n",
       "      <td>0.509803</td>\n",
       "      <td>-1.368419</td>\n",
       "      <td>0.509803</td>\n",
       "      <td>-1.368419</td>\n",
       "      <td>...</td>\n",
       "      <td>0.509803</td>\n",
       "      <td>0.509803</td>\n",
       "      <td>-1.368419</td>\n",
       "      <td>0.509803</td>\n",
       "      <td>0.509803</td>\n",
       "      <td>0.509803</td>\n",
       "      <td>0.509803</td>\n",
       "      <td>-1.368419</td>\n",
       "      <td>0.509803</td>\n",
       "      <td>0.509803</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1_9944824_C_T_b38</th>\n",
       "      <td>0.488223</td>\n",
       "      <td>0.488223</td>\n",
       "      <td>0.488223</td>\n",
       "      <td>0.488223</td>\n",
       "      <td>0.488223</td>\n",
       "      <td>0.488223</td>\n",
       "      <td>0.488223</td>\n",
       "      <td>-1.647751</td>\n",
       "      <td>0.488223</td>\n",
       "      <td>-1.647751</td>\n",
       "      <td>...</td>\n",
       "      <td>0.488223</td>\n",
       "      <td>0.488223</td>\n",
       "      <td>-1.647751</td>\n",
       "      <td>0.488223</td>\n",
       "      <td>0.488223</td>\n",
       "      <td>0.488223</td>\n",
       "      <td>0.488223</td>\n",
       "      <td>-1.647751</td>\n",
       "      <td>0.488223</td>\n",
       "      <td>0.488223</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr20_14358157_G_T_b38</th>\n",
       "      <td>1.024173</td>\n",
       "      <td>-0.438931</td>\n",
       "      <td>-0.438931</td>\n",
       "      <td>1.024173</td>\n",
       "      <td>-0.438931</td>\n",
       "      <td>-0.438931</td>\n",
       "      <td>-1.902035</td>\n",
       "      <td>-1.902035</td>\n",
       "      <td>-0.438931</td>\n",
       "      <td>1.024173</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.438931</td>\n",
       "      <td>1.024173</td>\n",
       "      <td>1.024173</td>\n",
       "      <td>1.024173</td>\n",
       "      <td>1.024173</td>\n",
       "      <td>-0.438931</td>\n",
       "      <td>-0.438931</td>\n",
       "      <td>-0.438931</td>\n",
       "      <td>-1.902035</td>\n",
       "      <td>-0.438931</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr20_14360685_A_T_b38</th>\n",
       "      <td>1.024173</td>\n",
       "      <td>-0.438931</td>\n",
       "      <td>-0.438931</td>\n",
       "      <td>1.024173</td>\n",
       "      <td>-0.438931</td>\n",
       "      <td>-0.438931</td>\n",
       "      <td>-1.902035</td>\n",
       "      <td>-1.902035</td>\n",
       "      <td>-0.438931</td>\n",
       "      <td>1.024173</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.438931</td>\n",
       "      <td>1.024173</td>\n",
       "      <td>1.024173</td>\n",
       "      <td>1.024173</td>\n",
       "      <td>1.024173</td>\n",
       "      <td>-0.438931</td>\n",
       "      <td>-0.438931</td>\n",
       "      <td>-0.438931</td>\n",
       "      <td>-1.902035</td>\n",
       "      <td>-0.438931</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr22_36992814_G_A_b38</th>\n",
       "      <td>1.709592</td>\n",
       "      <td>-0.466252</td>\n",
       "      <td>-0.466252</td>\n",
       "      <td>-0.466252</td>\n",
       "      <td>-0.466252</td>\n",
       "      <td>-0.466252</td>\n",
       "      <td>-0.466252</td>\n",
       "      <td>-0.466252</td>\n",
       "      <td>-0.466252</td>\n",
       "      <td>-0.466252</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.466252</td>\n",
       "      <td>1.709592</td>\n",
       "      <td>-0.466252</td>\n",
       "      <td>-0.466252</td>\n",
       "      <td>-0.466252</td>\n",
       "      <td>-0.466252</td>\n",
       "      <td>-0.466252</td>\n",
       "      <td>1.709592</td>\n",
       "      <td>1.709592</td>\n",
       "      <td>-0.466252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr22_46543361_T_C_b38</th>\n",
       "      <td>-0.096275</td>\n",
       "      <td>1.401330</td>\n",
       "      <td>-0.096275</td>\n",
       "      <td>-1.593879</td>\n",
       "      <td>-1.593879</td>\n",
       "      <td>1.401330</td>\n",
       "      <td>-0.096275</td>\n",
       "      <td>-0.096275</td>\n",
       "      <td>-1.593879</td>\n",
       "      <td>1.401330</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.593879</td>\n",
       "      <td>-1.593879</td>\n",
       "      <td>-0.096275</td>\n",
       "      <td>-0.096275</td>\n",
       "      <td>-0.096275</td>\n",
       "      <td>-1.593879</td>\n",
       "      <td>1.401330</td>\n",
       "      <td>-1.593879</td>\n",
       "      <td>-0.096275</td>\n",
       "      <td>1.401330</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr22_46572159_T_C_b38</th>\n",
       "      <td>-1.060016</td>\n",
       "      <td>0.727963</td>\n",
       "      <td>-1.060016</td>\n",
       "      <td>0.727963</td>\n",
       "      <td>-1.060016</td>\n",
       "      <td>0.727963</td>\n",
       "      <td>0.727963</td>\n",
       "      <td>0.727963</td>\n",
       "      <td>-1.060016</td>\n",
       "      <td>0.727963</td>\n",
       "      <td>...</td>\n",
       "      <td>-2.847995</td>\n",
       "      <td>-1.060016</td>\n",
       "      <td>0.727963</td>\n",
       "      <td>-1.060016</td>\n",
       "      <td>-1.060016</td>\n",
       "      <td>-2.847995</td>\n",
       "      <td>-1.060016</td>\n",
       "      <td>-1.060016</td>\n",
       "      <td>-1.060016</td>\n",
       "      <td>-1.060016</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>149 rows × 140 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        GTEX-113JC  GTEX-11DXX  GTEX-11EM3  GTEX-11EMC  \\\n",
       "snp                                                                      \n",
       "chr1_9919625_C_T_b38      0.531717    0.531717    0.531717    0.531717   \n",
       "chr1_9932244_A_C_b38      0.520278    0.520278    0.520278    0.520278   \n",
       "chr1_9934281_A_G_b38      0.531717    0.531717    0.531717    0.531717   \n",
       "chr1_9940342_T_A_b38      0.509803    0.509803    0.509803    0.509803   \n",
       "chr1_9944824_C_T_b38      0.488223    0.488223    0.488223    0.488223   \n",
       "...                            ...         ...         ...         ...   \n",
       "chr20_14358157_G_T_b38    1.024173   -0.438931   -0.438931    1.024173   \n",
       "chr20_14360685_A_T_b38    1.024173   -0.438931   -0.438931    1.024173   \n",
       "chr22_36992814_G_A_b38    1.709592   -0.466252   -0.466252   -0.466252   \n",
       "chr22_46543361_T_C_b38   -0.096275    1.401330   -0.096275   -1.593879   \n",
       "chr22_46572159_T_C_b38   -1.060016    0.727963   -1.060016    0.727963   \n",
       "\n",
       "                        GTEX-11GSP  GTEX-11I78  GTEX-11P81  GTEX-11TTK  \\\n",
       "snp                                                                      \n",
       "chr1_9919625_C_T_b38      0.531717    0.531717    0.531717   -1.536070   \n",
       "chr1_9932244_A_C_b38      0.520278    0.520278    0.520278   -1.347385   \n",
       "chr1_9934281_A_G_b38      0.531717    0.531717    0.531717   -1.536070   \n",
       "chr1_9940342_T_A_b38      0.509803    0.509803    0.509803   -1.368419   \n",
       "chr1_9944824_C_T_b38      0.488223    0.488223    0.488223   -1.647751   \n",
       "...                            ...         ...         ...         ...   \n",
       "chr20_14358157_G_T_b38   -0.438931   -0.438931   -1.902035   -1.902035   \n",
       "chr20_14360685_A_T_b38   -0.438931   -0.438931   -1.902035   -1.902035   \n",
       "chr22_36992814_G_A_b38   -0.466252   -0.466252   -0.466252   -0.466252   \n",
       "chr22_46543361_T_C_b38   -1.593879    1.401330   -0.096275   -0.096275   \n",
       "chr22_46572159_T_C_b38   -1.060016    0.727963    0.727963    0.727963   \n",
       "\n",
       "                        GTEX-11VI4  GTEX-11XUK  ...  GTEX-ZP4G  GTEX-ZQG8  \\\n",
       "snp                                             ...                         \n",
       "chr1_9919625_C_T_b38      0.531717   -1.536070  ...   0.531717   0.531717   \n",
       "chr1_9932244_A_C_b38      0.520278   -1.347385  ...   0.520278   0.520278   \n",
       "chr1_9934281_A_G_b38      0.531717   -1.536070  ...   0.531717   0.531717   \n",
       "chr1_9940342_T_A_b38      0.509803   -1.368419  ...   0.509803   0.509803   \n",
       "chr1_9944824_C_T_b38      0.488223   -1.647751  ...   0.488223   0.488223   \n",
       "...                            ...         ...  ...        ...        ...   \n",
       "chr20_14358157_G_T_b38   -0.438931    1.024173  ...  -0.438931   1.024173   \n",
       "chr20_14360685_A_T_b38   -0.438931    1.024173  ...  -0.438931   1.024173   \n",
       "chr22_36992814_G_A_b38   -0.466252   -0.466252  ...  -0.466252   1.709592   \n",
       "chr22_46543361_T_C_b38   -1.593879    1.401330  ...  -1.593879  -1.593879   \n",
       "chr22_46572159_T_C_b38   -1.060016    0.727963  ...  -2.847995  -1.060016   \n",
       "\n",
       "                        GTEX-ZTPG  GTEX-ZV6S  GTEX-ZVT2  GTEX-ZVT3  GTEX-ZXES  \\\n",
       "snp                                                                             \n",
       "chr1_9919625_C_T_b38    -1.536070   0.531717   0.531717   0.531717   0.531717   \n",
       "chr1_9932244_A_C_b38    -1.347385   0.520278   0.520278   0.520278   0.520278   \n",
       "chr1_9934281_A_G_b38    -1.536070   0.531717   0.531717   0.531717   0.531717   \n",
       "chr1_9940342_T_A_b38    -1.368419   0.509803   0.509803   0.509803   0.509803   \n",
       "chr1_9944824_C_T_b38    -1.647751   0.488223   0.488223   0.488223   0.488223   \n",
       "...                           ...        ...        ...        ...        ...   \n",
       "chr20_14358157_G_T_b38   1.024173   1.024173   1.024173  -0.438931  -0.438931   \n",
       "chr20_14360685_A_T_b38   1.024173   1.024173   1.024173  -0.438931  -0.438931   \n",
       "chr22_36992814_G_A_b38  -0.466252  -0.466252  -0.466252  -0.466252  -0.466252   \n",
       "chr22_46543361_T_C_b38  -0.096275  -0.096275  -0.096275  -1.593879   1.401330   \n",
       "chr22_46572159_T_C_b38   0.727963  -1.060016  -1.060016  -2.847995  -1.060016   \n",
       "\n",
       "                        GTEX-ZYVF  GTEX-ZYY3  GTEX-ZZPU  \n",
       "snp                                                      \n",
       "chr1_9919625_C_T_b38    -1.536070   0.531717   0.531717  \n",
       "chr1_9932244_A_C_b38    -1.347385   0.520278   0.520278  \n",
       "chr1_9934281_A_G_b38    -1.536070   0.531717   0.531717  \n",
       "chr1_9940342_T_A_b38    -1.368419   0.509803   0.509803  \n",
       "chr1_9944824_C_T_b38    -1.647751   0.488223   0.488223  \n",
       "...                           ...        ...        ...  \n",
       "chr20_14358157_G_T_b38  -0.438931  -1.902035  -0.438931  \n",
       "chr20_14360685_A_T_b38  -0.438931  -1.902035  -0.438931  \n",
       "chr22_36992814_G_A_b38   1.709592   1.709592  -0.466252  \n",
       "chr22_46543361_T_C_b38  -1.593879  -0.096275   1.401330  \n",
       "chr22_46572159_T_C_b38  -1.060016  -1.060016  -1.060016  \n",
       "\n",
       "[149 rows x 140 columns]"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotype_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "221888d6-6cba-4e1e-9b94-b2691e075e97",
   "metadata": {},
   "outputs": [],
   "source": [
    "plink_prefix_path = \"../../data/trans_qtl_calling/gtex/genotypes_filtered/plink.Vagina.GOBP_TISSUE_DEVELOPMENT\"\n",
    "proportions_loc = \"../../data/trans_qtl_calling/gtex/celltype_proportions/proportions-Vagina.txt\"\n",
    "covariates_file = \"../../data/trans_qtl_calling/gtex/covariates/Vagina.GOBP_TISSUE_DEVELOPMENT.supervised_surrogate_variables.txt\"\n",
    "other_covariates_file = \""
   ]
  },
  {
   "cell_type": "raw",
   "id": "1db75480-d392-4569-825b-6bcddc2b859b",
   "metadata": {},
   "source": [
    "plink_prefix_path = snakemake.params['plink_prefix']\n",
    "proportions_loc = snakemake.input['ctprops']\n",
    "covariates_file = snakemake.input['cov']\n",
    "output_loc = snakemake.output['output_loc']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5a2e16b9-66ba-4e62-8f67-981fd06e09df",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Mapping files: 100%|██████████| 3/3 [00:00<00:00, 33.45it/s]\n"
     ]
    }
   ],
   "source": [
    "# Load genotypes\n",
    "pr = genotypeio.PlinkReader(plink_prefix_path)\n",
    "genotype_df = pr.load_genotypes()\n",
    "variant_df = pr.bim.set_index('snp')[['chrom', 'pos']]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "8f33dca2-67e3-45fb-9599-930422f09792",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load covariates\n",
    "covariates_df = pd.read_csv(covariates_file, sep='\\t', index_col=0)\n",
    "\n",
    "# Filter to the donors for whom we have expression + genotype data\n",
    "donors_inc = genotype_df.columns.intersection(covariates_df.columns)\n",
    "\n",
    "# Load phenotypes, filter to the donors included and cell types with median score of at least 0.1\n",
    "phenotype_df = pd.read_csv(proportions_loc, sep=\"\\t\").set_index('cell_type')\n",
    "phenotype_df = phenotype_df.loc[phenotype_df.median(axis=1) > 0.1, donors_inc]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "60774bab-efae-48d7-b8b5-8ac74b81722c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GTEX-113JC</th>\n",
       "      <th>GTEX-11DXX</th>\n",
       "      <th>GTEX-11EM3</th>\n",
       "      <th>GTEX-11EMC</th>\n",
       "      <th>GTEX-11GSP</th>\n",
       "      <th>GTEX-11I78</th>\n",
       "      <th>GTEX-11P81</th>\n",
       "      <th>GTEX-11TTK</th>\n",
       "      <th>GTEX-11VI4</th>\n",
       "      <th>GTEX-11XUK</th>\n",
       "      <th>...</th>\n",
       "      <th>GTEX-ZP4G</th>\n",
       "      <th>GTEX-ZQG8</th>\n",
       "      <th>GTEX-ZTPG</th>\n",
       "      <th>GTEX-ZV6S</th>\n",
       "      <th>GTEX-ZVT2</th>\n",
       "      <th>GTEX-ZVT3</th>\n",
       "      <th>GTEX-ZXES</th>\n",
       "      <th>GTEX-ZYVF</th>\n",
       "      <th>GTEX-ZYY3</th>\n",
       "      <th>GTEX-ZZPU</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_type</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Epithelial_cells</th>\n",
       "      <td>0.7930</td>\n",
       "      <td>0.9287</td>\n",
       "      <td>0.0413</td>\n",
       "      <td>0.8718</td>\n",
       "      <td>0.0339</td>\n",
       "      <td>1.0438</td>\n",
       "      <td>0.8742</td>\n",
       "      <td>0.9934</td>\n",
       "      <td>0.9706</td>\n",
       "      <td>0.9993</td>\n",
       "      <td>...</td>\n",
       "      <td>0.8869</td>\n",
       "      <td>0.0659</td>\n",
       "      <td>0.8879</td>\n",
       "      <td>0.0806</td>\n",
       "      <td>0.8388</td>\n",
       "      <td>0.1999</td>\n",
       "      <td>0.9556</td>\n",
       "      <td>0.3423</td>\n",
       "      <td>0.8462</td>\n",
       "      <td>1.0362</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Keratinocytes</th>\n",
       "      <td>0.3034</td>\n",
       "      <td>0.3138</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.3832</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.4367</td>\n",
       "      <td>0.3861</td>\n",
       "      <td>0.4040</td>\n",
       "      <td>0.3983</td>\n",
       "      <td>0.4101</td>\n",
       "      <td>...</td>\n",
       "      <td>0.3736</td>\n",
       "      <td>0.0041</td>\n",
       "      <td>0.3334</td>\n",
       "      <td>0.0013</td>\n",
       "      <td>0.3402</td>\n",
       "      <td>0.0407</td>\n",
       "      <td>0.4022</td>\n",
       "      <td>0.0785</td>\n",
       "      <td>0.3192</td>\n",
       "      <td>0.4406</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 140 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  GTEX-113JC  GTEX-11DXX  GTEX-11EM3  GTEX-11EMC  GTEX-11GSP  \\\n",
       "cell_type                                                                      \n",
       "Epithelial_cells      0.7930      0.9287      0.0413      0.8718      0.0339   \n",
       "Keratinocytes         0.3034      0.3138      0.0000      0.3832      0.0000   \n",
       "\n",
       "                  GTEX-11I78  GTEX-11P81  GTEX-11TTK  GTEX-11VI4  GTEX-11XUK  \\\n",
       "cell_type                                                                      \n",
       "Epithelial_cells      1.0438      0.8742      0.9934      0.9706      0.9993   \n",
       "Keratinocytes         0.4367      0.3861      0.4040      0.3983      0.4101   \n",
       "\n",
       "                  ...  GTEX-ZP4G  GTEX-ZQG8  GTEX-ZTPG  GTEX-ZV6S  GTEX-ZVT2  \\\n",
       "cell_type         ...                                                          \n",
       "Epithelial_cells  ...     0.8869     0.0659     0.8879     0.0806     0.8388   \n",
       "Keratinocytes     ...     0.3736     0.0041     0.3334     0.0013     0.3402   \n",
       "\n",
       "                  GTEX-ZVT3  GTEX-ZXES  GTEX-ZYVF  GTEX-ZYY3  GTEX-ZZPU  \n",
       "cell_type                                                                \n",
       "Epithelial_cells     0.1999     0.9556     0.3423     0.8462     1.0362  \n",
       "Keratinocytes        0.0407     0.4022     0.0785     0.3192     0.4406  \n",
       "\n",
       "[2 rows x 140 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phenotype_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "a602a24c-cc66-46e7-b6d7-2da36bae844a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jpopp/.conda/envs/tensorqtl/lib/python3.11/site-packages/sklearn/preprocessing/_data.py:2627: UserWarning: n_quantiles (1000) is greater than the total number of samples (140). n_quantiles is set to n_samples.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "# Quantile normalize the phenotypes\n",
    "phenotype_df_norm = quantile_transform(phenotype_df, axis=1, output_distribution='normal')\n",
    "phenotype_df_norm = pd.DataFrame(phenotype_df_norm, index=phenotype_df.index, columns=phenotype_df.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "70db3555-897d-40c4-b76e-abdd28d3d89a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SV</th>\n",
       "      <th>GTEX-113JC</th>\n",
       "      <th>GTEX-11DXX</th>\n",
       "      <th>GTEX-11EM3</th>\n",
       "      <th>GTEX-11EMC</th>\n",
       "      <th>GTEX-11GSP</th>\n",
       "      <th>GTEX-11I78</th>\n",
       "      <th>GTEX-11P81</th>\n",
       "      <th>GTEX-11TTK</th>\n",
       "      <th>GTEX-11VI4</th>\n",
       "      <th>...</th>\n",
       "      <th>GTEX-ZP4G</th>\n",
       "      <th>GTEX-ZQG8</th>\n",
       "      <th>GTEX-ZTPG</th>\n",
       "      <th>GTEX-ZV6S</th>\n",
       "      <th>GTEX-ZVT2</th>\n",
       "      <th>GTEX-ZVT3</th>\n",
       "      <th>GTEX-ZXES</th>\n",
       "      <th>GTEX-ZYVF</th>\n",
       "      <th>GTEX-ZYY3</th>\n",
       "      <th>GTEX-ZZPU</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNP</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1_9919625_C_T_b38</th>\n",
       "      <td>SV1</td>\n",
       "      <td>0.038587</td>\n",
       "      <td>-0.039670</td>\n",
       "      <td>0.105410</td>\n",
       "      <td>-0.053653</td>\n",
       "      <td>0.132816</td>\n",
       "      <td>-0.156442</td>\n",
       "      <td>-0.056175</td>\n",
       "      <td>-0.061968</td>\n",
       "      <td>-0.073499</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.052557</td>\n",
       "      <td>0.091624</td>\n",
       "      <td>-0.043430</td>\n",
       "      <td>0.088830</td>\n",
       "      <td>-0.060455</td>\n",
       "      <td>0.131354</td>\n",
       "      <td>-0.095262</td>\n",
       "      <td>0.089424</td>\n",
       "      <td>0.021594</td>\n",
       "      <td>-0.193784</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1_9919625_C_T_b38</th>\n",
       "      <td>SV2</td>\n",
       "      <td>0.138770</td>\n",
       "      <td>-0.050425</td>\n",
       "      <td>-0.134885</td>\n",
       "      <td>0.059502</td>\n",
       "      <td>0.050024</td>\n",
       "      <td>0.000083</td>\n",
       "      <td>0.082671</td>\n",
       "      <td>0.220010</td>\n",
       "      <td>-0.051829</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.081541</td>\n",
       "      <td>-0.170614</td>\n",
       "      <td>-0.049160</td>\n",
       "      <td>-0.094879</td>\n",
       "      <td>-0.071630</td>\n",
       "      <td>0.121075</td>\n",
       "      <td>-0.029765</td>\n",
       "      <td>0.036436</td>\n",
       "      <td>0.105361</td>\n",
       "      <td>-0.029743</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1_9919625_C_T_b38</th>\n",
       "      <td>SV3</td>\n",
       "      <td>-0.090111</td>\n",
       "      <td>0.125861</td>\n",
       "      <td>-0.050761</td>\n",
       "      <td>-0.036838</td>\n",
       "      <td>-0.044150</td>\n",
       "      <td>-0.035638</td>\n",
       "      <td>0.021395</td>\n",
       "      <td>-0.058289</td>\n",
       "      <td>0.062501</td>\n",
       "      <td>...</td>\n",
       "      <td>0.014255</td>\n",
       "      <td>-0.043997</td>\n",
       "      <td>-0.041007</td>\n",
       "      <td>-0.035784</td>\n",
       "      <td>0.090860</td>\n",
       "      <td>-0.006237</td>\n",
       "      <td>-0.008421</td>\n",
       "      <td>-0.153538</td>\n",
       "      <td>0.004387</td>\n",
       "      <td>0.027481</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1_9919625_C_T_b38</th>\n",
       "      <td>SV4</td>\n",
       "      <td>0.028608</td>\n",
       "      <td>0.130073</td>\n",
       "      <td>-0.085788</td>\n",
       "      <td>0.031414</td>\n",
       "      <td>-0.087025</td>\n",
       "      <td>-0.054375</td>\n",
       "      <td>-0.019131</td>\n",
       "      <td>0.071823</td>\n",
       "      <td>-0.020164</td>\n",
       "      <td>...</td>\n",
       "      <td>0.052177</td>\n",
       "      <td>0.076382</td>\n",
       "      <td>-0.047688</td>\n",
       "      <td>0.038705</td>\n",
       "      <td>0.031387</td>\n",
       "      <td>0.049561</td>\n",
       "      <td>-0.195894</td>\n",
       "      <td>-0.028858</td>\n",
       "      <td>-0.024465</td>\n",
       "      <td>-0.039247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1_9919625_C_T_b38</th>\n",
       "      <td>SV5</td>\n",
       "      <td>-0.003912</td>\n",
       "      <td>0.140435</td>\n",
       "      <td>-0.089580</td>\n",
       "      <td>0.071391</td>\n",
       "      <td>0.083646</td>\n",
       "      <td>-0.088467</td>\n",
       "      <td>-0.061425</td>\n",
       "      <td>-0.121552</td>\n",
       "      <td>-0.030867</td>\n",
       "      <td>...</td>\n",
       "      <td>0.019518</td>\n",
       "      <td>-0.116345</td>\n",
       "      <td>0.152404</td>\n",
       "      <td>0.002726</td>\n",
       "      <td>0.069022</td>\n",
       "      <td>-0.036733</td>\n",
       "      <td>-0.145103</td>\n",
       "      <td>0.162525</td>\n",
       "      <td>0.006609</td>\n",
       "      <td>-0.061220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr22_46572159_T_C_b38</th>\n",
       "      <td>SV17</td>\n",
       "      <td>-0.092581</td>\n",
       "      <td>-0.024967</td>\n",
       "      <td>0.089723</td>\n",
       "      <td>0.060059</td>\n",
       "      <td>0.004937</td>\n",
       "      <td>-0.042914</td>\n",
       "      <td>-0.009192</td>\n",
       "      <td>-0.015913</td>\n",
       "      <td>-0.080479</td>\n",
       "      <td>...</td>\n",
       "      <td>0.027472</td>\n",
       "      <td>-0.038047</td>\n",
       "      <td>0.045054</td>\n",
       "      <td>-0.039964</td>\n",
       "      <td>-0.083676</td>\n",
       "      <td>-0.129791</td>\n",
       "      <td>-0.073323</td>\n",
       "      <td>0.053011</td>\n",
       "      <td>0.057591</td>\n",
       "      <td>-0.251469</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr22_46572159_T_C_b38</th>\n",
       "      <td>SV18</td>\n",
       "      <td>0.033971</td>\n",
       "      <td>0.153969</td>\n",
       "      <td>-0.056824</td>\n",
       "      <td>-0.054989</td>\n",
       "      <td>-0.188374</td>\n",
       "      <td>-0.025801</td>\n",
       "      <td>0.024387</td>\n",
       "      <td>-0.194183</td>\n",
       "      <td>-0.055419</td>\n",
       "      <td>...</td>\n",
       "      <td>0.004834</td>\n",
       "      <td>-0.050135</td>\n",
       "      <td>-0.118679</td>\n",
       "      <td>0.102540</td>\n",
       "      <td>0.115837</td>\n",
       "      <td>-0.070005</td>\n",
       "      <td>0.030836</td>\n",
       "      <td>-0.000598</td>\n",
       "      <td>-0.001644</td>\n",
       "      <td>-0.040801</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr22_46572159_T_C_b38</th>\n",
       "      <td>SV19</td>\n",
       "      <td>-0.097723</td>\n",
       "      <td>0.000990</td>\n",
       "      <td>0.051396</td>\n",
       "      <td>-0.187566</td>\n",
       "      <td>0.067813</td>\n",
       "      <td>0.126685</td>\n",
       "      <td>-0.013566</td>\n",
       "      <td>0.116059</td>\n",
       "      <td>-0.050961</td>\n",
       "      <td>...</td>\n",
       "      <td>0.016721</td>\n",
       "      <td>0.019348</td>\n",
       "      <td>-0.028593</td>\n",
       "      <td>-0.006873</td>\n",
       "      <td>-0.087005</td>\n",
       "      <td>-0.076171</td>\n",
       "      <td>-0.037504</td>\n",
       "      <td>0.054156</td>\n",
       "      <td>-0.054959</td>\n",
       "      <td>-0.066190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr22_46572159_T_C_b38</th>\n",
       "      <td>SV20</td>\n",
       "      <td>-0.045642</td>\n",
       "      <td>-0.052407</td>\n",
       "      <td>0.075512</td>\n",
       "      <td>0.072122</td>\n",
       "      <td>-0.153264</td>\n",
       "      <td>-0.056789</td>\n",
       "      <td>-0.027060</td>\n",
       "      <td>0.116842</td>\n",
       "      <td>0.047655</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.039039</td>\n",
       "      <td>-0.000896</td>\n",
       "      <td>0.060139</td>\n",
       "      <td>0.026911</td>\n",
       "      <td>0.035450</td>\n",
       "      <td>-0.084432</td>\n",
       "      <td>-0.011720</td>\n",
       "      <td>-0.042957</td>\n",
       "      <td>0.041175</td>\n",
       "      <td>-0.233448</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr22_46572159_T_C_b38</th>\n",
       "      <td>SV21</td>\n",
       "      <td>-0.089697</td>\n",
       "      <td>0.061013</td>\n",
       "      <td>0.125754</td>\n",
       "      <td>0.030319</td>\n",
       "      <td>-0.101690</td>\n",
       "      <td>-0.081098</td>\n",
       "      <td>0.054003</td>\n",
       "      <td>0.287459</td>\n",
       "      <td>0.154281</td>\n",
       "      <td>...</td>\n",
       "      <td>0.032125</td>\n",
       "      <td>0.028371</td>\n",
       "      <td>-0.043041</td>\n",
       "      <td>-0.017451</td>\n",
       "      <td>0.032883</td>\n",
       "      <td>0.007991</td>\n",
       "      <td>-0.065934</td>\n",
       "      <td>-0.023940</td>\n",
       "      <td>0.085814</td>\n",
       "      <td>-0.129893</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3129 rows × 141 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          SV  GTEX-113JC  GTEX-11DXX  GTEX-11EM3  GTEX-11EMC  \\\n",
       "SNP                                                                            \n",
       "chr1_9919625_C_T_b38     SV1    0.038587   -0.039670    0.105410   -0.053653   \n",
       "chr1_9919625_C_T_b38     SV2    0.138770   -0.050425   -0.134885    0.059502   \n",
       "chr1_9919625_C_T_b38     SV3   -0.090111    0.125861   -0.050761   -0.036838   \n",
       "chr1_9919625_C_T_b38     SV4    0.028608    0.130073   -0.085788    0.031414   \n",
       "chr1_9919625_C_T_b38     SV5   -0.003912    0.140435   -0.089580    0.071391   \n",
       "...                      ...         ...         ...         ...         ...   \n",
       "chr22_46572159_T_C_b38  SV17   -0.092581   -0.024967    0.089723    0.060059   \n",
       "chr22_46572159_T_C_b38  SV18    0.033971    0.153969   -0.056824   -0.054989   \n",
       "chr22_46572159_T_C_b38  SV19   -0.097723    0.000990    0.051396   -0.187566   \n",
       "chr22_46572159_T_C_b38  SV20   -0.045642   -0.052407    0.075512    0.072122   \n",
       "chr22_46572159_T_C_b38  SV21   -0.089697    0.061013    0.125754    0.030319   \n",
       "\n",
       "                        GTEX-11GSP  GTEX-11I78  GTEX-11P81  GTEX-11TTK  \\\n",
       "SNP                                                                      \n",
       "chr1_9919625_C_T_b38      0.132816   -0.156442   -0.056175   -0.061968   \n",
       "chr1_9919625_C_T_b38      0.050024    0.000083    0.082671    0.220010   \n",
       "chr1_9919625_C_T_b38     -0.044150   -0.035638    0.021395   -0.058289   \n",
       "chr1_9919625_C_T_b38     -0.087025   -0.054375   -0.019131    0.071823   \n",
       "chr1_9919625_C_T_b38      0.083646   -0.088467   -0.061425   -0.121552   \n",
       "...                            ...         ...         ...         ...   \n",
       "chr22_46572159_T_C_b38    0.004937   -0.042914   -0.009192   -0.015913   \n",
       "chr22_46572159_T_C_b38   -0.188374   -0.025801    0.024387   -0.194183   \n",
       "chr22_46572159_T_C_b38    0.067813    0.126685   -0.013566    0.116059   \n",
       "chr22_46572159_T_C_b38   -0.153264   -0.056789   -0.027060    0.116842   \n",
       "chr22_46572159_T_C_b38   -0.101690   -0.081098    0.054003    0.287459   \n",
       "\n",
       "                        GTEX-11VI4  ...  GTEX-ZP4G  GTEX-ZQG8  GTEX-ZTPG  \\\n",
       "SNP                                 ...                                    \n",
       "chr1_9919625_C_T_b38     -0.073499  ...  -0.052557   0.091624  -0.043430   \n",
       "chr1_9919625_C_T_b38     -0.051829  ...  -0.081541  -0.170614  -0.049160   \n",
       "chr1_9919625_C_T_b38      0.062501  ...   0.014255  -0.043997  -0.041007   \n",
       "chr1_9919625_C_T_b38     -0.020164  ...   0.052177   0.076382  -0.047688   \n",
       "chr1_9919625_C_T_b38     -0.030867  ...   0.019518  -0.116345   0.152404   \n",
       "...                            ...  ...        ...        ...        ...   \n",
       "chr22_46572159_T_C_b38   -0.080479  ...   0.027472  -0.038047   0.045054   \n",
       "chr22_46572159_T_C_b38   -0.055419  ...   0.004834  -0.050135  -0.118679   \n",
       "chr22_46572159_T_C_b38   -0.050961  ...   0.016721   0.019348  -0.028593   \n",
       "chr22_46572159_T_C_b38    0.047655  ...  -0.039039  -0.000896   0.060139   \n",
       "chr22_46572159_T_C_b38    0.154281  ...   0.032125   0.028371  -0.043041   \n",
       "\n",
       "                        GTEX-ZV6S  GTEX-ZVT2  GTEX-ZVT3  GTEX-ZXES  GTEX-ZYVF  \\\n",
       "SNP                                                                             \n",
       "chr1_9919625_C_T_b38     0.088830  -0.060455   0.131354  -0.095262   0.089424   \n",
       "chr1_9919625_C_T_b38    -0.094879  -0.071630   0.121075  -0.029765   0.036436   \n",
       "chr1_9919625_C_T_b38    -0.035784   0.090860  -0.006237  -0.008421  -0.153538   \n",
       "chr1_9919625_C_T_b38     0.038705   0.031387   0.049561  -0.195894  -0.028858   \n",
       "chr1_9919625_C_T_b38     0.002726   0.069022  -0.036733  -0.145103   0.162525   \n",
       "...                           ...        ...        ...        ...        ...   \n",
       "chr22_46572159_T_C_b38  -0.039964  -0.083676  -0.129791  -0.073323   0.053011   \n",
       "chr22_46572159_T_C_b38   0.102540   0.115837  -0.070005   0.030836  -0.000598   \n",
       "chr22_46572159_T_C_b38  -0.006873  -0.087005  -0.076171  -0.037504   0.054156   \n",
       "chr22_46572159_T_C_b38   0.026911   0.035450  -0.084432  -0.011720  -0.042957   \n",
       "chr22_46572159_T_C_b38  -0.017451   0.032883   0.007991  -0.065934  -0.023940   \n",
       "\n",
       "                        GTEX-ZYY3  GTEX-ZZPU  \n",
       "SNP                                           \n",
       "chr1_9919625_C_T_b38     0.021594  -0.193784  \n",
       "chr1_9919625_C_T_b38     0.105361  -0.029743  \n",
       "chr1_9919625_C_T_b38     0.004387   0.027481  \n",
       "chr1_9919625_C_T_b38    -0.024465  -0.039247  \n",
       "chr1_9919625_C_T_b38     0.006609  -0.061220  \n",
       "...                           ...        ...  \n",
       "chr22_46572159_T_C_b38   0.057591  -0.251469  \n",
       "chr22_46572159_T_C_b38  -0.001644  -0.040801  \n",
       "chr22_46572159_T_C_b38  -0.054959  -0.066190  \n",
       "chr22_46572159_T_C_b38   0.041175  -0.233448  \n",
       "chr22_46572159_T_C_b38   0.085814  -0.129893  \n",
       "\n",
       "[3129 rows x 141 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "covariates_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b323f924-44af-4550-ab32-679435fa20ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "# assert np.all(phenotype_df.columns==covariates_df.index)\n",
    "# assert covariates_df.index.isin(genotype_df.columns).all()\n",
    "\n",
    "# Perform cell type proportion QTL calling\n",
    "trans_df = trans.map_trans(genotype_df, phenotype_df_norm, covariates_df,\n",
    "                           pval_threshold=1, maf_threshold=0.05,\n",
    "                           batch_size=20000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "466a87fa-b360-4a31-9449-d724b5a5a8ed",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>iid</th>\n",
       "      <th>GTEX-113JC</th>\n",
       "      <th>GTEX-11DXX</th>\n",
       "      <th>GTEX-11EM3</th>\n",
       "      <th>GTEX-11EMC</th>\n",
       "      <th>GTEX-11GSP</th>\n",
       "      <th>GTEX-11I78</th>\n",
       "      <th>GTEX-11P81</th>\n",
       "      <th>GTEX-11TTK</th>\n",
       "      <th>GTEX-11VI4</th>\n",
       "      <th>GTEX-11XUK</th>\n",
       "      <th>...</th>\n",
       "      <th>GTEX-ZP4G</th>\n",
       "      <th>GTEX-ZQG8</th>\n",
       "      <th>GTEX-ZTPG</th>\n",
       "      <th>GTEX-ZV6S</th>\n",
       "      <th>GTEX-ZVT2</th>\n",
       "      <th>GTEX-ZVT3</th>\n",
       "      <th>GTEX-ZXES</th>\n",
       "      <th>GTEX-ZYVF</th>\n",
       "      <th>GTEX-ZYY3</th>\n",
       "      <th>GTEX-ZZPU</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>snp</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1_9919625_C_T_b38</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1_9932244_A_C_b38</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1_9934281_A_G_b38</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1_9940342_T_A_b38</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1_9944824_C_T_b38</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr20_14358157_G_T_b38</th>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr20_14360685_A_T_b38</th>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr22_36992814_G_A_b38</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr22_46543361_T_C_b38</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr22_46572159_T_C_b38</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>149 rows × 140 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "iid                     GTEX-113JC  GTEX-11DXX  GTEX-11EM3  GTEX-11EMC  \\\n",
       "snp                                                                      \n",
       "chr1_9919625_C_T_b38             2           2           2           2   \n",
       "chr1_9932244_A_C_b38             2           2           2           2   \n",
       "chr1_9934281_A_G_b38             2           2           2           2   \n",
       "chr1_9940342_T_A_b38             2           2           2           2   \n",
       "chr1_9944824_C_T_b38             2           2           2           2   \n",
       "...                            ...         ...         ...         ...   \n",
       "chr20_14358157_G_T_b38           2           1           1           2   \n",
       "chr20_14360685_A_T_b38           2           1           1           2   \n",
       "chr22_36992814_G_A_b38           1           0           0           0   \n",
       "chr22_46543361_T_C_b38           1           2           1           0   \n",
       "chr22_46572159_T_C_b38           1           2           1           2   \n",
       "\n",
       "iid                     GTEX-11GSP  GTEX-11I78  GTEX-11P81  GTEX-11TTK  \\\n",
       "snp                                                                      \n",
       "chr1_9919625_C_T_b38             2           2           2           1   \n",
       "chr1_9932244_A_C_b38             2           2           2           1   \n",
       "chr1_9934281_A_G_b38             2           2           2           1   \n",
       "chr1_9940342_T_A_b38             2           2           2           1   \n",
       "chr1_9944824_C_T_b38             2           2           2           1   \n",
       "...                            ...         ...         ...         ...   \n",
       "chr20_14358157_G_T_b38           1           1           0           0   \n",
       "chr20_14360685_A_T_b38           1           1           0           0   \n",
       "chr22_36992814_G_A_b38           0           0           0           0   \n",
       "chr22_46543361_T_C_b38           0           2           1           1   \n",
       "chr22_46572159_T_C_b38           1           2           2           2   \n",
       "\n",
       "iid                     GTEX-11VI4  GTEX-11XUK  ...  GTEX-ZP4G   GTEX-ZQG8   \\\n",
       "snp                                             ...                           \n",
       "chr1_9919625_C_T_b38             2           1  ...           2           2   \n",
       "chr1_9932244_A_C_b38             2           1  ...           2           2   \n",
       "chr1_9934281_A_G_b38             2           1  ...           2           2   \n",
       "chr1_9940342_T_A_b38             2           1  ...           2           2   \n",
       "chr1_9944824_C_T_b38             2           1  ...           2           2   \n",
       "...                            ...         ...  ...         ...         ...   \n",
       "chr20_14358157_G_T_b38           1           2  ...           1           2   \n",
       "chr20_14360685_A_T_b38           1           2  ...           1           2   \n",
       "chr22_36992814_G_A_b38           0           0  ...           0           1   \n",
       "chr22_46543361_T_C_b38           0           2  ...           0           0   \n",
       "chr22_46572159_T_C_b38           1           2  ...           0           1   \n",
       "\n",
       "iid                     GTEX-ZTPG   GTEX-ZV6S   GTEX-ZVT2   GTEX-ZVT3   \\\n",
       "snp                                                                      \n",
       "chr1_9919625_C_T_b38             1           2           2           2   \n",
       "chr1_9932244_A_C_b38             1           2           2           2   \n",
       "chr1_9934281_A_G_b38             1           2           2           2   \n",
       "chr1_9940342_T_A_b38             1           2           2           2   \n",
       "chr1_9944824_C_T_b38             1           2           2           2   \n",
       "...                            ...         ...         ...         ...   \n",
       "chr20_14358157_G_T_b38           2           2           2           1   \n",
       "chr20_14360685_A_T_b38           2           2           2           1   \n",
       "chr22_36992814_G_A_b38           0           0           0           0   \n",
       "chr22_46543361_T_C_b38           1           1           1           0   \n",
       "chr22_46572159_T_C_b38           2           1           1           0   \n",
       "\n",
       "iid                     GTEX-ZXES   GTEX-ZYVF   GTEX-ZYY3   GTEX-ZZPU   \n",
       "snp                                                                     \n",
       "chr1_9919625_C_T_b38             2           1           2           2  \n",
       "chr1_9932244_A_C_b38             2           1           2           2  \n",
       "chr1_9934281_A_G_b38             2           1           2           2  \n",
       "chr1_9940342_T_A_b38             2           1           2           2  \n",
       "chr1_9944824_C_T_b38             2           1           2           2  \n",
       "...                            ...         ...         ...         ...  \n",
       "chr20_14358157_G_T_b38           1           1           0           1  \n",
       "chr20_14360685_A_T_b38           1           1           0           1  \n",
       "chr22_36992814_G_A_b38           0           1           1           0  \n",
       "chr22_46543361_T_C_b38           2           0           1           2  \n",
       "chr22_46572159_T_C_b38           1           1           1           1  \n",
       "\n",
       "[149 rows x 140 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotype_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "4d6ebf4e-616e-4043-ad20-0b941b49ae7a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['SV1', 'SV2', 'SV3', 'SV4', 'SV5', 'SV6', 'SV7', 'SV8', 'SV9', 'SV10',\n",
       "       'SV11', 'SV12', 'SV13', 'SV14', 'SV15', 'SV16', 'SV17', 'SV18', 'SV19',\n",
       "       'SV20', 'SV21'],\n",
       "      dtype='object', name='SV')"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cov_df.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "683b2166-ad03-4d5e-92a0-5c4487d36906",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>phenotype_id</th>\n",
       "      <th>pval</th>\n",
       "      <th>b</th>\n",
       "      <th>b_se</th>\n",
       "      <th>af</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1_9919625_C_T_b38</td>\n",
       "      <td>Epithelial_cells</td>\n",
       "      <td>0.340832</td>\n",
       "      <td>0.087145</td>\n",
       "      <td>0.091116</td>\n",
       "      <td>0.871429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1_9919625_C_T_b38</td>\n",
       "      <td>Keratinocytes</td>\n",
       "      <td>0.017541</td>\n",
       "      <td>0.442265</td>\n",
       "      <td>0.183564</td>\n",
       "      <td>0.871429</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             variant_id      phenotype_id      pval         b      b_se  \\\n",
       "0  chr1_9919625_C_T_b38  Epithelial_cells  0.340832  0.087145  0.091116   \n",
       "1  chr1_9919625_C_T_b38     Keratinocytes  0.017541  0.442265  0.183564   \n",
       "\n",
       "         af  \n",
       "0  0.871429  \n",
       "1  0.871429  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "s = 'chr1_9919625_C_T_b38'\n",
    "geno_df = genotype_df.loc[genotype_df.index == s, :]\n",
    "cov_df = covariates_df.loc[covariates_df.index == s, :].set_index('SV').T\n",
    "trans.map_trans(geno_df, phenotype_df_norm, cov_df,\n",
    "                pval_threshold=1, maf_threshold=0.05,\n",
    "                batch_size=20000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "c2e810f2-0b93-48a2-bff2-d1f5daf47308",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n",
      "trans-QTL mapping\n",
      "  * 140 samples\n",
      "  * 2 phenotypes\n",
      "  * 21 covariates\n",
      "  * 1 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.00 min\n",
      "  * 1 variants passed MAF >= 0.05 filtering\n",
      "done.\n"
     ]
    }
   ],
   "source": [
    "qtl_all = []\n",
    "for s in covariates_df.index.unique().tolist():\n",
    "    geno_df = genotype_df.loc[genotype_df.index == s, :]\n",
    "    cov_df = covariates_df.loc[covariates_df.index == s, :].set_index('SV').T\n",
    "    qtl_all.append(trans.map_trans(geno_df, phenotype_df_norm, cov_df,\n",
    "                                   pval_threshold=1, maf_threshold=0.05))\n",
    "ct_qtls = pd.concat(qtl_all)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "c1499ed0-9278-403f-a504-6885c5ab1c55",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>phenotype_id</th>\n",
       "      <th>pval</th>\n",
       "      <th>b</th>\n",
       "      <th>b_se</th>\n",
       "      <th>af</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1_9919625_C_T_b38</td>\n",
       "      <td>Epithelial_cells</td>\n",
       "      <td>0.340832</td>\n",
       "      <td>0.087145</td>\n",
       "      <td>0.091116</td>\n",
       "      <td>0.871429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1_9919625_C_T_b38</td>\n",
       "      <td>Keratinocytes</td>\n",
       "      <td>0.017541</td>\n",
       "      <td>0.442265</td>\n",
       "      <td>0.183564</td>\n",
       "      <td>0.871429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1_9932244_A_C_b38</td>\n",
       "      <td>Epithelial_cells</td>\n",
       "      <td>0.364105</td>\n",
       "      <td>0.083052</td>\n",
       "      <td>0.091154</td>\n",
       "      <td>0.870504</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1_9932244_A_C_b38</td>\n",
       "      <td>Keratinocytes</td>\n",
       "      <td>0.020111</td>\n",
       "      <td>0.431131</td>\n",
       "      <td>0.182955</td>\n",
       "      <td>0.870504</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1_9934281_A_G_b38</td>\n",
       "      <td>Epithelial_cells</td>\n",
       "      <td>0.340832</td>\n",
       "      <td>0.087145</td>\n",
       "      <td>0.091116</td>\n",
       "      <td>0.871429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr22_36992814_G_A_b38</td>\n",
       "      <td>Keratinocytes</td>\n",
       "      <td>0.156314</td>\n",
       "      <td>0.304195</td>\n",
       "      <td>0.213208</td>\n",
       "      <td>0.115942</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr22_46543361_T_C_b38</td>\n",
       "      <td>Epithelial_cells</td>\n",
       "      <td>0.377322</td>\n",
       "      <td>-0.059597</td>\n",
       "      <td>0.067249</td>\n",
       "      <td>0.539568</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr22_46543361_T_C_b38</td>\n",
       "      <td>Keratinocytes</td>\n",
       "      <td>0.725437</td>\n",
       "      <td>-0.048548</td>\n",
       "      <td>0.137902</td>\n",
       "      <td>0.539568</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr22_46572159_T_C_b38</td>\n",
       "      <td>Epithelial_cells</td>\n",
       "      <td>0.164198</td>\n",
       "      <td>0.113429</td>\n",
       "      <td>0.081028</td>\n",
       "      <td>0.796429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr22_46572159_T_C_b38</td>\n",
       "      <td>Keratinocytes</td>\n",
       "      <td>0.819997</td>\n",
       "      <td>-0.038177</td>\n",
       "      <td>0.167398</td>\n",
       "      <td>0.796429</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>298 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                variant_id      phenotype_id      pval         b      b_se  \\\n",
       "0     chr1_9919625_C_T_b38  Epithelial_cells  0.340832  0.087145  0.091116   \n",
       "1     chr1_9919625_C_T_b38     Keratinocytes  0.017541  0.442265  0.183564   \n",
       "0     chr1_9932244_A_C_b38  Epithelial_cells  0.364105  0.083052  0.091154   \n",
       "1     chr1_9932244_A_C_b38     Keratinocytes  0.020111  0.431131  0.182955   \n",
       "0     chr1_9934281_A_G_b38  Epithelial_cells  0.340832  0.087145  0.091116   \n",
       "..                     ...               ...       ...       ...       ...   \n",
       "1   chr22_36992814_G_A_b38     Keratinocytes  0.156314  0.304195  0.213208   \n",
       "0   chr22_46543361_T_C_b38  Epithelial_cells  0.377322 -0.059597  0.067249   \n",
       "1   chr22_46543361_T_C_b38     Keratinocytes  0.725437 -0.048548  0.137902   \n",
       "0   chr22_46572159_T_C_b38  Epithelial_cells  0.164198  0.113429  0.081028   \n",
       "1   chr22_46572159_T_C_b38     Keratinocytes  0.819997 -0.038177  0.167398   \n",
       "\n",
       "          af  \n",
       "0   0.871429  \n",
       "1   0.871429  \n",
       "0   0.870504  \n",
       "1   0.870504  \n",
       "0   0.871429  \n",
       "..       ...  \n",
       "1   0.115942  \n",
       "0   0.539568  \n",
       "1   0.539568  \n",
       "0   0.796429  \n",
       "1   0.796429  \n",
       "\n",
       "[298 rows x 6 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trans_df = pd.concat(qtl_all)\n",
    "trans_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f51d9d6-bb80-42c7-a20d-39a161fe08f0",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "tensorqtl",
   "language": "python",
   "name": "tensorqtl"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
