```{r}
library(tidyverse)
library(vroom)
library(qvalue)
library(MatrixGenerics)
library(RColorBrewer)
```

```{r}
classify_dynamics <- function(b_g, b_i, b_gi, tmin, tmax, gmin, gmax, threshold=1) {
  early_effect = (gmin*b_g + tmin*b_i + gmin*tmin*b_gi) - (gmax*b_g + tmin*b_i + gmax*tmin*b_gi)
  late_effect = (gmin*b_g + tmax*b_i + gmin*tmax*b_gi) - (gmax*b_g + tmax*b_i + gmax*tmax*b_gi)
  
  if (sign(early_effect) == sign(late_effect)) {
    if (abs(early_effect) >= abs(late_effect)) {
      class = "early"
    } else {
      class = "late"
    }
    # QTL switching signs
  } else {
    if ((abs(early_effect) >= abs(late_effect)) & (abs(late_effect) < threshold)) {
      class = "early"
    } else if ((abs(early_effect) < abs(late_effect)) & (abs(early_effect) < threshold)) {
      class = "late"
      # Switch magnitude is greater than or equal to threshold - is this switching behavior significant enough
    } else if ((abs(early_effect) >= threshold) & (abs(late_effect) >= threshold)) {
      class = "switch"
    } 
    else {
      class = "unassigned"
    }
  }
  class
}
```


Load the cardiomyocyte trajectory data
```{r}
cm_genotypes <- vroom("data/dynamic_qtl_calling/eb-cm_15binstrimmed/pseudobulk_tmm/nipals/10clpcs/genotype_df.tsv") %>%
  column_to_rownames("snp")
cm_gmins <- rowMins(as.matrix(cm_genotypes))
cm_gmaxs <- rowMaxs(as.matrix(cm_genotypes))
cm_granges <- tibble(variant_id=rownames(cm_genotypes), gmin=cm_gmins, gmax=cm_gmaxs) 
cm_all <- vroom("results/dynamic_qtl_calling/eb-cm_15binstrimmed/pseudobulk_tmm/nipals/10clpcs/tensorqtl_interactions.all.tsv")
cm_emt <- vroom("results/dynamic_qtl_calling/eb-cm_15binstrimmed/pseudobulk_tmm/nipals/10clpcs/tensorqtl_interactions.cis_qtl_top_assoc.txt.gz")
cm_pseudotime <- vroom("data/dynamic_qtl_calling/eb-cm_15binstrimmed/pseudobulk_tmm/pseudotime.tsv")
cm_hits <- filter(cm_emt, pval_adj_bh <= 0.1) %>%
  left_join(cm_granges, by="variant_id") %>%
  mutate(tmin=min(cm_pseudotime$pseudotime), tmax=max(cm_pseudotime$pseudotime)) %>%
  mutate(class=pmap_chr(.l=list(b_g, b_i, b_gi, tmin, tmax, gmin, gmax), .f=classify_dynamics))

neur_genotypes <- vroom("data/dynamic_qtl_calling/eb-neur_15binstrimmed/pseudobulk_tmm/nipals/10clpcs/genotype_df.tsv") %>%
  column_to_rownames("snp")
neur_gmins <- rowMins(as.matrix(neur_genotypes))
neur_gmaxs <- rowMaxs(as.matrix(neur_genotypes))
neur_granges <- tibble(variant_id=rownames(neur_genotypes), gmin=neur_gmins, gmax=neur_gmaxs) 
neur_all <- vroom("results/dynamic_qtl_calling/eb-neur_15binstrimmed/pseudobulk_tmm/nipals/10clpcs/tensorqtl_interactions.all.tsv")
neur_emt <- vroom("results/dynamic_qtl_calling/eb-neur_15binstrimmed/pseudobulk_tmm/nipals/10clpcs/tensorqtl_interactions.cis_qtl_top_assoc.txt.gz")
neur_pseudotime <- vroom("data/dynamic_qtl_calling/eb-neur_15binstrimmed/pseudobulk_tmm/pseudotime.tsv")
neur_hits <- filter(neur_emt, pval_adj_bh <= 0.1) %>%
  left_join(neur_granges, by="variant_id") %>%
  mutate(tmin=min(neur_pseudotime$pseudotime), tmax=max(neur_pseudotime$pseudotime)) %>%
  mutate(class=pmap_chr(.l=list(b_g, b_i, b_gi, tmin, tmax, gmin, gmax), .f=classify_dynamics))

hep_genotypes <- vroom("data/dynamic_qtl_calling/eb-hep_15binstrimmed/pseudobulk_tmm/nipals/10clpcs/genotype_df.tsv") %>%
  column_to_rownames("snp")
hep_gmins <- rowMins(as.matrix(hep_genotypes))
hep_gmaxs <- rowMaxs(as.matrix(hep_genotypes))
hep_granges <- tibble(variant_id=rownames(hep_genotypes), gmin=hep_gmins, gmax=hep_gmaxs) 
hep_all <- vroom("results/dynamic_qtl_calling/eb-hep_15binstrimmed/pseudobulk_tmm/nipals/10clpcs/tensorqtl_interactions.all.tsv")
hep_emt <- vroom("results/dynamic_qtl_calling/eb-hep_15binstrimmed/pseudobulk_tmm/nipals/10clpcs/tensorqtl_interactions.cis_qtl_top_assoc.txt.gz")
hep_pseudotime <- vroom("data/dynamic_qtl_calling/eb-hep_15binstrimmed/pseudobulk_tmm/pseudotime.tsv")
hep_hits <- filter(hep_emt, pval_adj_bh <= 0.1) %>%
  left_join(hep_granges, by="variant_id") %>%
  mutate(tmin=min(hep_pseudotime$pseudotime), tmax=max(hep_pseudotime$pseudotime)) %>%
  mutate(class=pmap_chr(.l=list(b_g, b_i, b_gi, tmin, tmax, gmin, gmax), .f=classify_dynamics))

```

```{r}
ggplot(count(cm_hits_classified, qtl_class), aes(x=qtl_class, y=n)) +
  geom_col() + ggtitle("Cardiomyocyte Dynamic eQTLs by Type")
ggplot(count(neur_hits_classified, qtl_class), aes(x=qtl_class, y=n)) +
  geom_col() + ggtitle("Neuronal Dynamic eQTLs by Type")
head(cm_hits_classified, n=10)
head(neur_hits_classified, n=10)
```

Visualize some examples
```{r}
cm_exp <- vroom("data/dynamic_qtl_calling/eb-cm_15binstrimmed/pseudobulk_tmm/expression.bed.gz") %>%
  filter(gene %in% c(cm_hits$phenotype_id)) %>%
  select(-c(`#chr`, start, end)) %>%
  pivot_longer(!gene, names_to="ind_type", values_to="expression")
cm_genotype_df <- vroom("data/dynamic_qtl_calling/eb-cm_15binstrimmed/pseudobulk_tmm/nipals/10clpcs/genotype_df.tsv") %>% 
  filter(snp %in% c(cm_hits$variant_id)) %>%
  pivot_longer(!snp, names_to="ind_type", values_to="genotype")
cm_pseudotime <- vroom("data/dynamic_qtl_calling/eb-cm_15binstrimmed/pseudobulk_tmm/pseudotime.tsv") %>%
  mutate(pseudotime_bin=as.numeric(str_extract(ind_type, "[^bin]+$")))
```

Visualize an early CM example
```{r}
g <- "RILPL1"
v <- "rs112931584"
select(filter(cm_exp, gene==g), c("ind_type", "expression")) %>%
  inner_join(filter(cm_genotype_df, snp==v), c("ind_type", "expression"), by="ind_type") %>%
  inner_join(cm_pseudotime, by="ind_type") %>%
  mutate(genotype=factor(genotype), pseudotime_bin=factor(pseudotime_bin)) %>%
  ggplot(aes(x=pseudotime_bin, y=expression, fill=genotype)) +
  geom_boxplot(width=0.8, lwd=1.25) +
  theme_classic(base_size=15) +
  ylab("RILPL1 Expression") +
  xlab("Pseudotime Bin") +
  scale_fill_manual(values=brewer.pal(3, "Set1"))
```

Visualize a late CM example
```{r}
g <- "TXNL4B"
v <- "rs8047930"
select(filter(cm_exp, gene==g), c("ind_type", "expression")) %>%
  inner_join(filter(cm_genotype_df, snp==v), c("ind_type", "expression"), by="ind_type") %>%
  inner_join(cm_pseudotime, by="ind_type") %>%
  mutate(genotype=factor(genotype), pseudotime_bin=factor(pseudotime_bin)) %>%
  ggplot(aes(x=pseudotime_bin, y=expression, fill=genotype)) +
  geom_boxplot()
```

## Replication Between Lineages
High replication of dynamic eQTLs across the two trajectories 
```{r}
cm_all_rep_in_neur <- cm_hits_classified %>%
  select(c(phenotype_id, variant_id)) %>%
  inner_join(neur_all, by=c("phenotype_id", "variant_id"))
hist(qvalue(cm_all_rep_in_neur$pval_gi))
```

Especially high among early tests
```{r}
cm_early_rep_in_neur <- cm_hits_classified %>%
  filter(qtl_class=="early") %>%
  select(c(phenotype_id, variant_id)) %>%
  inner_join(neur_all, by=c("phenotype_id", "variant_id"))
hist(qvalue(cm_early_rep_in_neur$pval_gi, lambda=0.5))

neur_early_rep_in_cm <- neur_hits_classified %>%
  filter(qtl_class=="early") %>%
  select(c(phenotype_id, variant_id)) %>%
  inner_join(cm_all, by=c("phenotype_id", "variant_id"))
hist(qvalue(neur_early_rep_in_cm$pval_gi, lambda=0.5))
```

```{r}
cm_late_rep_in_neur <- cm_hits_classified %>%
  filter(qtl_class=="late") %>%
  select(c(phenotype_id, variant_id)) %>%
  inner_join(neur_all, by=c("phenotype_id", "variant_id"))
hist(qvalue(cm_late_rep_in_neur$pval_gi, lambda=0.5))

neur_late_rep_in_cm <- neur_hits_classified %>%
  filter(qtl_class=="late") %>%
  select(c(phenotype_id, variant_id)) %>%
  inner_join(cm_all, by=c("phenotype_id", "variant_id"))
hist(qvalue(neur_late_rep_in_cm$pval_gi, lambda=0.5))
```

Visualize a trait example
```{r fig.height=3, fig.width=8}
g <- "COL1A2"
v <- "rs10487252"

cm_exp <- vroom("data/dynamic_qtl_calling/eb-cm_15binstrimmed/pseudobulk_tmm/expression.bed.gz") %>%
  filter(gene ==g) %>%
  dplyr::select(-c(`#chr`, start, end)) %>%
  pivot_longer(!gene, names_to="ind_type", values_to="expression")
cm_pseudotime <- vroom("data/dynamic_qtl_calling/eb-cm_15binstrimmed/pseudobulk_tmm/pseudotime.tsv") %>%
  mutate(pseudotime_bin=as.numeric(str_extract(ind_type, "[^bin]+$")))


cm_genotype_df_full <- vroom("data/dynamic_qtl_calling/eb-cm_15binstrimmed/pseudobulk_tmm/nipals/10clpcs/genotype_df.tsv") 
cm_genotype_df <- cm_genotype_df_full %>% 
  filter(snp ==v) %>%
  pivot_longer(!snp, names_to="ind_type", values_to="genotype")

qtl_df <- dplyr::select(filter(cm_exp, gene==g), c("ind_type", "expression")) %>%
  inner_join(filter(cm_genotype_df, snp==v), c("ind_type", "expression"), by="ind_type") %>%
  inner_join(cm_pseudotime, by="ind_type") %>%
  mutate(genotype=factor(genotype), pseudotime_bin=factor(pseudotime_bin)) 

ggplot(qtl_df, aes(x=pseudotime_bin, y=expression, fill=genotype)) +
  geom_boxplot(linewidth=1.25) +
  theme_classic() +
  xlab("") + ylab("") +
  theme(axis.text.x = element_blank(), 
        axis.ticks.x = element_blank(),
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank()) +
  labs(legend.position="none") +
  scale_fill_manual(values=c("#B51700", "#0076BA"))
```

Now focus on residual expression
```{r}
cm_covariates_df <- vroom("data/dynamic_qtl_calling/eb-cm_15binstrimmed/pseudobulk_tmm/nipals/10clpcs/covariate_df.tsv")
full_qtl_df <- filter(cm_exp, gene == g) %>%
  select(-c(gene)) %>%
  inner_join(select(cm_genotype_df, !snp), by="ind_type") %>% 
  inner_join(select(cm_pseudotime, !pseudotime_bin), by="ind_type") %>%
  inner_join(cm_covariates_df, by=c("ind_type" = "sample")) %>%
  column_to_rownames("ind_type")
```

```{r}
qtl_m <- lm("expression ~ .", full_qtl_df)
exp_residuals <- qtl_m$residuals

qtl_df$residual_expression <- exp_residuals
ggplot(qtl_df, aes(x=pseudotime_bin, y=residual_expression, fill=genotype)) +
  geom_boxplot()
```

Diagnostics - look at the residuals plot
```{r}
qtl_m_full <- lm("expression ~ . + genotype * pseudotime", full_qtl_df)

plot(qtl_m_full)
```

