---
title: "Cell-ID Evaluation (EB Side)"
output: html_notebook
---

```{r}
library(tidyverse)
library(Matrix.utils)
library(fgsea)
library(susieR)
library(pathways)
library(vroom)
library(gseasusie)
library(nullranges)
set.seed(1)
```

```{r}
eb_sighits_loc <- "/project2/gilad/jpopp/ebQTL/results/static_qtl_calling/eb_cellid/pseudobulk_tmm/basic/8pcs/signif_variant_gene_pairs.tsv"
eb_sighits_bed_loc <- "/project2/gilad/jpopp/ebQTL/results/static_eqtl_followup/qtl_sets/tensorqtl/original/signif_variant_gene_pairs.bed"
eb_sighits_overlap_loc <- "/project2/gilad/jpopp/ebQTL/results/static_eqtl_followup/qtl_sets/tensorqtl/original/signif_variant_gene_pairs.all_tissue_overlap.bed"
gtf_loc <- "/project2/gilad/jpopp/ebQTL/data/gencode/gencode.hg38.filtered.gtf"
```

## Overview of eQTLs Identified
How many eQTLs do we find per cell type?
```{r}
eb_all <- vroom(eb_sighits_loc)

eb_all %>%
  dplyr::count(celltype) %>%
  ggplot(aes(x=celltype, y=n)) +
  geom_bar(stat="identity") +
  coord_flip() +
  theme_classic()
```

How many eGenes do we find per cell type?
```{r fig.height=4, fig.width=3}
eb_tophits_loc <- "/project2/gilad/jpopp/ebQTL/results/static_qtl_calling/eb_cellid/pseudobulk_tmm/basic/8pcs/tensorqtl_permutations.sighits.tsv"

eb_top_per_gene <- vroom(eb_tophits_loc)

egenes_per_celltype <- eb_top_per_gene %>%
  dplyr::count(context) 

ggplot(egenes_per_celltype, aes(x=context, y=n)) +
  geom_bar(stat="identity") +
  coord_flip() +
  theme_classic(base_size=15)
```


## Exploring GTEx Overlap
How many of the eQTLs that we identify have previously been identified as an eQTL?

To measure overlap, first, nest so that each QTL is only represented once
```{r}
eb_distinct <- vroom(eb_sighits_bed_loc) %>% 
  dplyr::select(c(EB_HGNC, EB_VARIANT_ID)) %>%
  distinct()

gtex_distinct <- vroom(eb_sighits_overlap_loc, col_names=c("CHR", "START", "STOP", "EB_ENSG", "EB_HGNC", 
                                                          "RSID", "CELLTYPE", "CHR1", "START1", 
                                                          "STOP1", "GTEX_ENSG", "GTEX_REF", "GTEX_ALT")) %>%
  dplyr::select(EB_HGNC, RSID) %>%
  distinct()

nrow(gtex_distinct) / nrow(eb_distinct)
```

## Gene set enrichment
Wrangle GO BP gene sets
```{r}
gmt_file <- "/project2/gilad/jpopp/ebQTL/data/gene_sets/c5.go.bp.v2022.1.Hs.symbols.gmt"

gmt_lines <- readLines(gmt_file)

gmt_list <- lapply(gmt_lines, function(x) unlist(strsplit(x, "\t")))
gmt_list2 <- lapply(gmt_list, function(x) list(x[1], x[2], x[3:length(x)]))

gmt_df <- as_tibble(do.call("rbind", gmt_list2)) %>%
  mutate(across(V1:V2, unlist)) %>%
  dplyr::rename(geneset=V1, link=V2, genes=V3) %>%
  filter(str_sub(geneset, 1, 4) == "GOBP") # filter to BP 

gmt_mat <- gmt_df %>%
  dplyr::select(-c(link)) %>%
  unnest(genes) %>%
  filter(genes %in% eb_distinct$EB_HGNC) %>%
  pivot_wider(names_from = genes, values_from = genes, values_fn = length, values_fill = 0) %>%
  column_to_rownames("geneset")

# subset to genesets with between 10 and 200 genes
valid_genesets <- (10 <= rowSums(gmt_mat)) & (rowSums(gmt_mat) < 200)
gmt_mat <- t(as.matrix(gmt_mat[which(valid_genesets),]))
```

# GSEA
### Pathway Analysis
```{r}
is_novel <- !rownames(gmt_mat) %in% gtex_distinct$EB_HGNC
ora <- gseasusie::fit_ora(X=gmt_mat, y=is_novel)
ora$bhHypergeometric <- p.adjust(ora$pHypergeometric, method="BH")
ora$bhFishersExact <- p.adjust(ora$pFishersExact, method="BH")
ora %>% arrange(pFishersExact) %>% 
  dplyr::select(geneSet, oddsRatio, bhFishersExact) %>%
  filter(bhFishersExact <= 0.05)
```

### Validation
We expect to see some of these due to the depletion of some of these gene sets in regular eQTLs

To validate, let's look at the distribution of TSS distance among the tissue development eQTLs
```{r}
tissue_dev_genes <- rownames(gmt_mat)[which(gmt_mat[,"GOBP_TISSUE_DEVELOPMENT"]==1)]
novel_egenes_tissue_dev <- intersect(tissue_dev_genes, setdiff(all_egenes, overlap_egenes))

novel_tissue_dev_qtls <- filter(eb_all, EB_HGNC %in% novel_egenes_tissue_dev) %>%
  unite(gv, EB_HGNC, EB_VARIANT_ID, sep="--", remove=F)
```

To do so, we also need the gene TSS
```{r}
gtf <- vroom("data/gencode/gencode.hg38.filtered.gtf") %>%
  mutate(tss = if_else(strand == "+", start, end)) %>%
  dplyr::select(hgnc, tss)
```

<!-- The connective tissue development enrichment is exciting, see which cell types that came from -->
<!-- ```{r} -->
<!-- connective_tissue_genes <- rownames(gmt_mat)[which(gmt_mat[,"GOBP_CONNECTIVE_TISSUE_DEVELOPMENT"]==1)] -->
<!-- novel_egenes_connective_tissue <- intersect(connective_tissue_genes, setdiff(all_egenes, overlap_egenes)) -->

<!-- novel_connective_tissue_qtls <- filter(eb_all, phenotype_id %in% novel_egenes_connective_tissue) -->
<!-- ``` -->

<!-- How about tissue development? -->
<!-- ```{r} -->
<!-- tissue_dev_genes <- rownames(gmt_mat)[which(gmt_mat[,"GOBP_TISSUE_DEVELOPMENT"]==1)] -->
<!-- novel_egenes_tissue_dev <- intersect(tissue_dev_genes, setdiff(all_egenes, overlap_egenes)) -->

<!-- novel_tissue_dev_qtls <- filter(eb_all, EB_HGNC %in% novel_egenes_tissue_dev) -->
<!-- ``` -->

<!-- ## Does this hold up when comparing to a background set? -->
<!-- ```{r} -->
<!-- matched_hits_loc <- "/project2/gilad/jpopp/ebQTL/results/static_qtl_calling/eb_cellid/pseudobulk_tmm/basic/8pcs/signif-matched_variant_gene_pairs.bed" -->
<!-- matched_hits_overlap_loc <- "results/static_eqtl_followup/eb_cellid/pseudobulk_tmm/basic/8pcs/signif-matched_variant_gene_pairs.full_gtex_overlap.bed" -->

<!-- matched_background_loc <- "/project2/gilad/jpopp/ebQTL/results/static_qtl_calling/eb_cellid/pseudobulk_tmm/basic/8pcs/background-matched_variant_gene_pairs.bed" -->
<!-- matched_background_overlap_loc <- "results/static_eqtl_followup/eb_cellid/pseudobulk_tmm/basic/8pcs/background-matched_variant_gene_pairs.full_gtex_overlap.bed" -->

<!-- matched_hits_overlap <- vroom(matched_hits_overlap_loc, col_names=c("CHR", "START", "STOP", "EB_ENSG", "EB_HGNC",  -->
<!--                                                           "RSID", "CELLTYPE", "CHR1", "START1",  -->
<!--                                                           "STOP1", "GTEX_ENSG", "GTEX_REF", "GTEX_ALT")) -->
<!-- hits_all <- vroom(matched_hits_loc) -->

<!-- matched_background_overlap <- vroom(matched_background_overlap_loc, col_names=c("CHR", "START", "STOP", "EB_ENSG", "EB_HGNC",  -->
<!--                                                           "RSID", "CELLTYPE", "CHR1", "START1",  -->
<!--                                                           "STOP1", "GTEX_ENSG", "GTEX_REF", "GTEX_ALT")) -->
<!-- background_all <- vroom(matched_background_loc) -->
<!-- ``` -->

<!-- Yes, it does hold up. 85% (24987 of 29466) of our EB eQTLs show up as an eQTL somewhere in GTEx,  -->
<!-- compared to 54% (16012 of 29466) of SNP-gene pairs matched for MAF and distance to TSS. -->

<!-- And we can assess whether this set was appropriately matched -->
<!-- ```{r} -->
<!-- matcher <- readRDS("results/static_qtl_calling/eb_cellid/pseudobulk_tmm/basic/8pcs/signif-hits-matcher.rds") -->
<!-- plotCovariate(matcher) -->
<!-- plotCovariate(matcher, "dist2tss") -->
<!-- ``` -->

