```{r}
library(tidyverse)
library(viridis)
```

```{r}
gtf_loc <- "data/gencode/gencode.hg38.filtered.gtf"
```

```{r}
gencode <- vroom(gtf_loc, col_select=c("seqname", "source", "feature", "start", "end", "score", "strand", "frame", "attribute", "hgnc", "ensg"))
```
Load all summary stats
```{r}
eb_sumstats_all <- vroom("results/static_qtl_calling/eb_cellid/pseudobulk_tmm/basic/8pcs/tensorqtl_nominal.all.tsv") %>%
  dplyr::rename(rsid=variant_id)
```

Load a variant dictionary to map rsID to chromosome position
```{r}
snp_map <- vroom("data/genotypes/human.YRI.hg38.all.AF.gencode.rsid_snpinfo_map.tsv",
                 col_names=c("rsid", "variant_id"))
```

## Visualize an example
```{r}
# gene <- "BCL11B" #SCZ
gene <- "FAM20C" #Cell Type Proportions
tss <- filter(gencode, hgnc==gene) %>%
  mutate(tss=if_else(strand=="+", start, end)) %>%
  pull(tss)

locus_df <- eb_sumstats_all %>%
  filter((phenotype_id==gene)) %>% 
  left_join(snp_map, by="rsid") %>%
  separate(variant_id, into=c("chr", "pos", "ref", "alt", "build"), sep="_") %>%
  mutate(nlog10p=-log10(pval_nominal)) %>%
  mutate(pos=as.numeric(pos))
```

```{r fig.width=5, fig.height=4}
ggplot(locus_df, aes(x=pos, y=nlog10p, color=nlog10p)) +
  geom_point() +
  geom_vline(xintercept=tss, linetype="dashed", color="red") +
  scale_color_viridis() +
  theme_classic() +
  facet_wrap(vars(celltype))
```

## Complex Trait Overlap
```{r}
scz_sumstats <- vroom("data/gwas/zhang/sumstats/PASS_Schizophrenia_Pardinas2018.sumstats")
```

```{r}
locus_pos <- filter(snp_map, rsid=="")
```


Map rsID to chromosome position
```{r}
scz_sumstats_locus <- scz_sumstats %>%
  left_join(snp_map, by=c("SNP"="rsid")) %>%
  separate(col=variant_id, into=c("chr", "pos", "ref", "alt", "build"), sep="_") %>%
  filter(chr=="chr14") %>%
  mutate(pos=as.numeric(pos)) %>%
  filter((pos >= tss-50000) & (pos <= tss+50000)) %>%
  mutate(p=2*pnorm(q=abs(Z), lower.tail=FALSE)) %>%
  mutate(nlog10p=-log10(p))
  
ggplot(um, aes(x=pos, y=nlog10p, color=nlog10p)) +
  geom_point() +
  geom_vline(xintercept=tss, linetype="dashed", color="red") +
  scale_color_viridis() +
  theme_classic()
```

Compare the two
```{r}
trait <- scz_sumstats_locus %>%
  select(c(pos, nlog10p)) %>%
  mutate(phenotype="SCZ")
qtl <- locus_df %>%
  filter(celltype=="PNS-glia") %>%
  select(c(pos, nlog10p)) %>%
  mutate(phenotype="expression")

bind_rows(trait, qtl) %>%
  ggplot(aes(x=pos, y=nlog10p, color=nlog10p)) +
  geom_point() +
  geom_vline(xintercept=tss, linetype="dashed", color="red") +
  scale_color_viridis(direction=-1) +
  facet_grid(rows=vars(phenotype)) 
```

## Dynamic eQTL
```{r}
hdl_sumstats <- fread("data/gwas/zhang/sumstats/UKB_460K.biochemistry_HDLcholesterol.sumstats")[, .(SNP, Z)]
hdl <- hdl[,P := z_to_p(Z)]
```

Load the dynamic eQTL data
```{r}
locus_gene <- "CCDC117"
dynamic_all <- vroom("results/dynamic_qtl_calling/eb-hep_15binstrimmed/pseudobulk_tmm/nipals/10clpcs/tensorqtl_interactions.all.tsv")
tmed_dynamic <- filter(dynamic_all, phenotype_id == locus_gene) %>%
  mutate(nlog10p = -log10(pval_gi))
bim_file <- vroom("data/dynamic_qtl_calling/eb-hep_15binstrimmed/pseudobulk_tmm/nipals/genotypes_filtered_plink.bim",
                  col_names=c("CHR", "RSID", "NA", "POS", "REF", "ALT"))
tmed_snps <- filter(bim_file, RSID %in% tmed_dynamic$variant_id) %>%
  select(CHR, POS, RSID)

dynamic_viz <- left_join(tmed_dynamic, tmed_snps, by=c("variant_id"="RSID"))
ggplot(dynamic_viz, aes(x=POS, y=nlog10p)) +
  geom_point()
```

Look at the subset that overlaps w GTEx
```{r}
overlap_dynamic_qtls <- vroom("results/static_eqtl_followup/eb_cellid/pseudobulk_tmm/basic/8pcs/dynamic_variant_gene_pairs.full_gtex_overlap.bed",
                              col_names=c("CHR", "START", "STOP", "EB_ENSG", "EB_HGNC", 
                                                          "RSID", "TRAJECTORY", "CHR1", "START1", 
                                                          "STOP1", "GTEX_ENSG", "GTEX_REF", "GTEX_ALT")) %>%
  filter(EB_HGNC==locus_gene, RSID %in% tmed_dynamic$variant_id)

dynamic_viz <- dynamic_viz %>%
  mutate(overlap = variant_id %in% overlap_dynamic_qtls$RSID)

ggplot(dynamic_viz, aes(x=POS, y=nlog10p)) +
  geom_point() +
  facet_grid(rows=vars(overlap))
```




## CRM Analysis
To understand the context of CRM loci, we can just get a quick view on the UMAP
```{r}
umap <- vroom("data/single_cell_objects/eb_pflog1ppfnorm.hvg.umap_embedding.txt", col_names=c("UMAP_1", "UMAP_2"))
celltypes <- vroom("data/fca/eb_cellid_labels.tsv")

sc_annotated <- celltypes %>%
  mutate(umap1=umap$UMAP_1, umap2=umap$UMAP_2)

pseudocell_map <- vroom("data/fast_topics/cell_pseudocell_mapping.tsv")

sc_annotated <- right_join(sc_annotated, pseudocell_map, by="cell")
```

```{r}
tmem <- vroom("results/cellregmap_eqtl_calling/eb_cellid/pseudobulk_tmm/basic/mash-signif.fasttopics_10topics.TMEM213.cellregmap.betas.tsv")

tmem %>%
  dplyr::select(PSEUDOCELL, BETA_GXC) %>%
  right_join(sc_annotated, by=c("PSEUDOCELL"="pseudocell_15")) %>%
  sample_n(100000) %>%
  ggplot(aes(x=umap1, y=umap2, color=BETA_GXC)) +
  geom_point(size=1e-5) +
  theme_classic() +
  scale_color_viridis()
```
```{r}
gata <- vroom("results/cellregmap_eqtl_calling/eb_cellid/pseudobulk_tmm/basic/mash-signif.fasttopics_10topics.GATA4.cellregmap.betas.tsv")

gata %>%
  dplyr::select(PSEUDOCELL, BETA_GXC) %>%
  right_join(sc_annotated, by=c("PSEUDOCELL"="pseudocell_15"), multiple="all") %>%
  sample_n(100000) %>%
  ggplot(aes(x=umap1, y=umap2, color=BETA_GXC)) +
  geom_point(size=1e-5) +
  theme_classic()
```
