```{r}
library(tidyverse)
library(viridis)
```

```{r}
gtf_loc <- "data/gencode/gencode.hg38.filtered.gtf"
```

```{r}
gencode <- vroom(gtf_loc, col_select=c("seqname", "source", "feature", "start", "end", "score", "strand", "frame", "attribute", "hgnc", "ensg"))
```
Load all summary stats
```{r}
eb_sumstats_all <- vroom("results/static_qtl_calling/eb_cellid/pseudobulk_tmm/basic/8pcs/tensorqtl_nominal.all.tsv") %>%
  dplyr::rename(rsid=variant_id)
```

Load a variant dictionary to map rsID to chromosome position
```{r}
snp_map <- vroom("data/genotypes/human.YRI.hg38.all.AF.gencode.rsid_snpinfo_map.tsv",
                 col_names=c("rsid", "variant_id"))
```

## Visualize an example
```{r}
# gene <- "BCL11B" #SCZ
gene <- "FAM20C" #Cell Type Proportions
tss <- filter(gencode, hgnc==gene) %>%
  mutate(tss=if_else(strand=="+", start, end)) %>%
  pull(tss)

locus_df <- eb_sumstats_all %>%
  filter((phenotype_id==gene)) %>% 
  left_join(snp_map, by="rsid") %>%
  separate(variant_id, into=c("chr", "pos", "ref", "alt", "build"), sep="_") %>%
  mutate(nlog10p=-log10(pval_nominal)) %>%
  mutate(pos=as.numeric(pos))
```

```{r fig.width=5, fig.height=4}
ggplot(locus_df, aes(x=pos, y=nlog10p, color=nlog10p)) +
  geom_point() +
  geom_vline(xintercept=tss, linetype="dashed", color="red") +
  scale_color_viridis() +
  theme_classic() +
  facet_wrap(vars(celltype))
```

## Complex Trait Overlap
```{r}
scz_sumstats <- vroom("data/gwas/zhang/sumstats/PASS_Schizophrenia_Pardinas2018.sumstats")
```

```{r}
locus_pos <- filter(snp_map, rsid=="")
```


Map rsID to chromosome position
```{r}
scz_sumstats_locus <- scz_sumstats %>%
  left_join(snp_map, by=c("SNP"="rsid")) %>%
  separate(col=variant_id, into=c("chr", "pos", "ref", "alt", "build"), sep="_") %>%
  filter(chr=="chr14") %>%
  mutate(pos=as.numeric(pos)) %>%
  filter((pos >= tss-50000) & (pos <= tss+50000)) %>%
  mutate(p=2*pnorm(q=abs(Z), lower.tail=FALSE)) %>%
  mutate(nlog10p=-log10(p))
  
ggplot(um, aes(x=pos, y=nlog10p, color=nlog10p)) +
  geom_point() +
  geom_vline(xintercept=tss, linetype="dashed", color="red") +
  scale_color_viridis() +
  theme_classic()
```

Compare the two
```{r}
trait <- scz_sumstats_locus %>%
  select(c(pos, nlog10p)) %>%
  mutate(phenotype="SCZ")
qtl <- locus_df %>%
  filter(celltype=="PNS-glia") %>%
  select(c(pos, nlog10p)) %>%
  mutate(phenotype="expression")

bind_rows(trait, qtl) %>%
  ggplot(aes(x=pos, y=nlog10p, color=nlog10p)) +
  geom_point() +
  geom_vline(xintercept=tss, linetype="dashed", color="red") +
  scale_color_viridis(direction=-1) +
  facet_grid(rows=vars(phenotype)) 
```

## eQTL Violin/ Box Plot
```{r}
exp <- vroom()
```




