```{r}
library(tidyverse)
library(vroom)
```

```{r}
trans_eqtls <- vroom("/project2/gilad/jpopp/ebQTL/data/trans_qtl_calling/gtex/trans_eqtl_results_2.tsv")
variant_groups <- vroom("results/static_eqtl_followup/eb_cellid/pseudobulk_tmm/basic/8pcs/trans_eqtl_variant_candidate_info.GOBP_CIRCULATORY_SYSTEM_DEVELOPMENT.Artery-Tibial.tsv")

trans_eqtls_grouped <- left_join(trans_eqtls, variant_groups, by="variant_id")
```

```{r}
qq_viz <- trans_eqtls_grouped %>%
  arrange(group, pval) %>% group_by(group) %>%
  add_tally() %>% mutate(group_index=row_number()) %>%
  mutate(nlog10_pval_unif=map2_dbl(group_index, n, function(a,b){-log10(a/b)})) %>%
  mutate(nlog10_pval=-log10(pval)) %>%
  mutate(group=factor(group, levels=c("pathway_novel",
                                      "pathway_matched",
                                      "ciseqtl_matched",
                                      "covariate_matched")))
```

View by group
```{r}
ggplot(qq_viz, aes(x = nlog10_pval_unif, y = nlog10_pval, color=group)) +
  geom_line() +
  geom_abline(intercept = 0, slope = 1, color = "red", linetype = "dashed") +
  labs(x = "Expected (-log10 p-values)",
       y = "Observed (-log10 p-values)",
       title = "QQ Plot") +
  theme_minimal()
```

The covariate hits here don't seem driven by cross-mappability, but they also don't reach genome-wide significance 

## Include LD blocks
First, I want to filter to PCGs 
```{r}
gtf_loc <- "/project2/gilad/kenneth/References/human/cellranger/cellranger4.0/refdata-gex-GRCh38-2020-A/genes/genes.gtf"
pull_gene_type <- function(attr) {
  # str_split(attr, "\"")[[1]][11] for the online version
  str_split(attr, "\"")[[1]][6] # for the kenneth version
}

pull_gene_name <- function(attr) {
  # str_split(attr, "\"")[[1]][15] for the online version
  str_split(attr, "\"")[[1]][8] # for the kenneth version
}

pull_gene_ensg <- function(attr) {
  # str_split(attr, "\"")[[1]][3] for the online version
  str_split(attr, "\"")[[1]][2] # for the kenneth version
}
gencode <- vroom(gtf_loc, col_names=c("seqname", "source", "feature", "start", "end", "score", "strand", "frame", "attribute"), skip=5) %>%
  filter(seqname %in% paste0("chr", seq(1, 22))) %>%
  filter(feature == "gene") %>%
  mutate(type=map_chr(attribute, pull_gene_type)) %>%
  mutate(hgnc=map_chr(attribute, pull_gene_name)) %>%
  mutate(ensg=map_chr(attribute, pull_gene_ensg)) %>%
  filter(type=="protein_coding")

pcgs <- pull(gencode, ensg)
```

Load the original variant info
```{r}
trans_candidates<- vroom("results/static_eqtl_followup/eb_cellid/pseudobulk_tmm/basic/8pcs/trans_eqtl_variant_candidate_info.GOBP_TISSUE_DEVELOPMENT.Artery-Tibial.tsv") %>%
  mutate(cis_gene=phenotype_id_ensg)
```


Load trans eQTL results and filter to PCGs
```{r}
trans_eqtls <- vroom("/project2/gilad/jpopp/ebQTL/data/trans_qtl_calling/gtex/trans_eqtl_results_with_ld.tsv")

qq_ld_viz <- trans_eqtls %>%
  filter(phenotype_id %in% pcgs) %>%
  arrange(pval) %>% 
  add_tally() %>% mutate(index=row_number()) %>%
  mutate(nlog10_pval_unif=map2_dbl(index, n, function(a,b){-log10(a/b)})) %>%
  mutate(nlog10_pval=-log10(pval))
```

```{r}
ggplot(qq_ld_viz, aes(x = nlog10_pval_unif, y = nlog10_pval)) +
  geom_line() +
  geom_abline(intercept = 0, slope = 1, color = "red", linetype = "dashed") +
  labs(x = "Expected (-log10 p-values)",
       y = "Observed (-log10 p-values)",
       title = "QQ Plot") +
  theme_minimal()
```

Hierarchical FDR control with TreeBH
```{r}
treebh_info <- trans_eqtls %>%
  filter(phenotype_id %in% pcgs) %>%
  left_join(select(trans_candidates, c(variant_id, cis_gene)), by="variant_id") %>%
  select(cis_gene, phenotype_id, variant_id, pval) %>%
  dplyr::rename(locus=cis_gene, gene=phenotype_id, variant=variant_id) %>%
  rowid_to_column("level3") %>%
  group_by(locus, gene) %>%
  mutate(level2=cur_group_id()) %>%
  ungroup %>% group_by(locus) %>%
  mutate(level1=cur_group_id()) %>%
  ungroup()

treebh_pvals <- as.vector(treebh_info$pval)
treebh_groups <- as.matrix(select(treebh_info, c(level1, level2, level3)))

treeBH_results <- get_TreeBH_selections(pvals = treebh_pvals,
                                        groups = treebh_groups,
                                        q = rep(0.2, 3))

```

# Include LD, with pathway focus
```{r}
trans_candidates_circ <- vroom("results/static_eqtl_followup/eb_cellid/pseudobulk_tmm/basic/8pcs/trans_eqtl_variant_candidate_info.GOBP_CIRCULATORY_SYSTEM_DEVELOPMENT.Artery-Tibial.tsv") %>%
  mutate(cis_gene=phenotype_id_ensg)

trans_eqtls_circ <- vroom("/project2/gilad/jpopp/ebQTL/data/trans_qtl_calling/gtex/trans_eqtl_results_with_ld.pathway.tsv")
```

```{r}
treebh_info_circ <- trans_eqtls_circ %>%
  filter(phenotype_id %in% pcgs) %>%
  left_join(select(trans_candidates_circ, c(variant_id, cis_gene)), by="variant_id") %>%
  select(cis_gene, phenotype_id, variant_id, pval) %>%
  dplyr::rename(locus=cis_gene, gene=phenotype_id, variant=variant_id) %>%
  rowid_to_column("level3") %>%
  group_by(locus, gene) %>%
  mutate(level2=cur_group_id()) %>%
  ungroup %>% group_by(locus) %>%
  mutate(level1=cur_group_id()) %>%
  ungroup()

treebh_pvals_circ <- as.vector(treebh_info_circ$pval)
treebh_groups_circ <- as.matrix(select(treebh_info_circ, c(level1, level2, level3)))

treeBH_results_circ <- get_TreeBH_selections(pvals = treebh_pvals_circ,
                                        groups = treebh_groups_circ,
                                        q = rep(0.5, 3))
```

