---
title: "Visualizing eQTLs"
output: html_notebook
---

```{r}
library(mashr)
library(corrplot)
library(tidyverse)
library(rmeta)
library(vroom)
library(RColorBrewer)
library(vcfR)
library(zellkonverter)
library(slingshot)
```

# Load Data
## Load single-cell data
```{r}
sc <- readH5AD("/project2/gilad/jpopp/ebQTL/data/single_cell_objects/highpass/cm_lineage2.h5ad")
```


## Load normalized pseudobulk data
```{r}
pseudobulk <- map_df(paste0("/project2/gilad/jpopp/ebQTL/data/static/cm_lineage2/pseudobulk_tmm/basic/", c("ipsc", "mesendo", "meso", "cp", "cm"), "/expression.tsv"), 
                    ~vroom(.x, id="path")) %>%
  mutate(type=map_chr(path, function(s){str_split(s, "/")[[1]][11]}), .keep="unused") %>%
  pivot_longer(!c(gene, type), names_to="ind", values_to="exp")
```


## Load genotype data
This needs to be changed to a more appropriate genotype file
```{r}
genotypes <- vroom("/project2/gilad/jpopp/ebQTL/data/static/highpass_cellid_all/pseudobulk-scran/basic/Erythroblasts/genotypes.tsv") %>%
  pivot_longer(!snp_id, names_to="ind", values_to="genotype")
```

## Load QTL data
```{r}
m <- readRDS("/project2/gilad/jpopp/ebQTL/results/static/cm_lineage2/pseudobulk_tmm/basic/4pcs/mashr.tophits.rds")
```

# Expression Preprocessing
## Load single-cell data and perform pseudotime smoothing per individual

# Visualize eQTLs in Pseudobulk
## Select eQTLs for visualization
Which of the eGenes that we found overlap with AF GWAS genes?
```{r}
af_genes <- read_tsv("/project2/gilad/jpopp/ebQTL/data/scDRS/gs_files/zhang.magma_10kb_1000.74_traits.gs") %>%
  filter(TRAIT=="PASS_AtrialFibrillation_Nielsen2018") %>%
  pull(GENESET) %>%
  str_split(",") %>% 
  unlist
```

Find any significant eQTLs whose effects are best explained by those signals, starting with component 1 (which is highest in IPSC, lower elsewhere)
```{r}
stage_colors <- brewer.pal(5, "Reds")
stage_levels <- c("ipsc", "mesendo", "meso", "cp", "cm")
```

```{r}
sighits <- get_significant_results(m)
egenes <- sapply(names(sighits), function(s){str_extract(s, "[^-]+")})
af_egenes <- intersect(af_genes, egenes)
length(af_egenes)
```

```{r}
component_per_hit <- as_tibble(m$posterior_weights) %>%
  mutate(gv=rownames(get_lfsr(m)), .before=1) %>%
  pivot_longer(!gv, names_to="component", values_to="weight") %>%
  mutate(component=str_extract(component, "[^.]+")) %>%
  group_by(gv, component) %>%
  summarize(contribution=sum(weight)) %>%
  slice_max(contribution) %>%
  arrange(desc(contribution)) %>%
  separate(gv, into=c("gene", "SNP"), sep="--")
```

```{r}
filter(component_per_hit, gene %in% af_egenes)
```

There are plenty of AF-associated genes with eQTLs, lets further filter to the ones displaying some meaningful dynamics
```{r}
filter(component_per_hit, gene %in% af_egenes) %>%
  filter(component=="ED_PCA_1")
```

```{r}
mash_plot_meta(m, "CAND2--chr3_12794264", colors=meta.colors(box=stage_colors))
```

```{r}
pseudobulk %>%
  filter(gene=="CAND2") %>%
  left_join(filter(genotypes, snp_id=="chr3_12794264"), by="ind") %>%
  mutate(genotype=factor(genotype, levels=c(0, 1, 2))) %>%
  mutate(type=factor(type, levels=stage_levels)) %>%
  ggplot(aes(x=type, y=exp, fill=genotype)) +
  geom_boxplot()
```

Which variants look to be closest to having a transient effect?
```{r}
as_tibble(m$posterior_weights) %>%
  mutate(gv=rownames(get_lfsr(m)), .before=1) %>%
  pivot_longer(!gv, names_to="component", values_to="weight") %>%
  mutate(component=str_extract(component, "[^.]+")) %>%
  group_by(gv, component) %>%
  summarize(contribution=sum(weight)) %>%
  filter(component=="ED_PCA_3") %>%
  arrange(desc(contribution)) %>%
  separate(gv, into=c("gene", "SNP"), sep="--") %>%
  filter(gene %in% af_egenes)
```

```{r}
mash_plot_meta(m, "FGF17--chr8_21992451", colors=meta.colors(box=stage_colors))
mash_plot_meta(m, "SPTB--chr14_64862469", colors=meta.colors(box=stage_colors))
```

```{r}
pseudobulk %>%
  filter(gene=="FGF17") %>%
  left_join(filter(genotypes, snp_id=="chr8_21992451"), by="ind") %>%
  mutate(genotype=factor(genotype, levels=c(0, 1, 2))) %>%
  mutate(type=factor(type, levels=stage_levels)) %>%
  ggplot(aes(x=type, y=exp, fill=genotype)) +
  geom_boxplot()

pseudobulk %>%
  filter(gene=="SPTB") %>%
  left_join(filter(genotypes, snp_id=="chr14_64862469"), by="ind") %>%
  mutate(genotype=factor(genotype, levels=c(0, 1, 2))) %>%
  mutate(type=factor(type, levels=stage_levels)) %>%
  drop_na() %>%
  ggplot(aes(x=type, y=exp, fill=genotype)) +
  geom_boxplot()
```

Alternatively, can look at mesoderm- or mesendoderm-specific significant hits
```{r}
meso_sp_qtls <- m$result$lfsr %>%
  as_tibble(rownames="gv") %>%
  pivot_longer(-gv, names_to="type", values_to="lfsr") %>%
  filter(lfsr<=0.1) %>%
  group_by(gv) %>%
  dplyr::select(-lfsr) %>%
  nest() %>%
  mutate(n_contexts=sapply(data, nrow)) %>%
  filter(n_contexts==1) %>%
  unnest(data) %>%
  filter(type %in% c("meso", "mesendo")) %>%
  separate(gv, into=c("gene", "SNPS"), sep="--", remove=FALSE)

meso_sp_qtls
```

```{r}
intersect(meso_sp_qtls$gene, af_egenes)
```

# Visualize eQTLs in single-cell data
```{r}
as_tibble(colData(sc), rownames="cell") %>%
  select(cell, donor_id, stage) %>%
  rename(ind=donor_id) %>%
  left_join(filter(genotypes, snp_id=="chr3_12794264"), by="ind") %>%
  mutate(genotype=factor(genotype)) %>%
  mutate("CAND2"=assays(sc)[['data']]["CAND2",]) %>%
  ggplot(aes(x=stage, y=CAND2, fill=genotype)) +
  geom_violin()
```
Not loving the look of diffusion pseudotime
```{r}
um <- slingshot(sc, clusterLabels="stage", reducedDim="X_pca")
```

