---
title: "Comparison of QTL Calling in IPSCs"
output: html_notebook
---

```{r, echo=FALSE, results=FALSE, warning=FALSE}
library(tidyverse)
library(qvalue)
```

How do our re-analyzed eQTLs from the variance QTL paper's data compare to theirs?

```{r, warning=FALSE, message=FALSE}
sarkar_tophits <- read_table("/project2/gilad/jpopp/data_from_abhishek/mean.txt.gz") %>%
  drop_na(var_chr, var_start, var_end)

# bonferroni correction
sarkar_tophits <- sarkar_tophits %>%
  mutate(p_bonf=if_else(p_nominal * num_vars < 1, p_nominal * num_vars, 1)) %>%
  mutate(q_bonf=qvalue(.$p_bonf)$q) %>%
  mutate(q_empirical=qvalue(.$p_empirical)$q) %>%
  mutate(q_beta=qvalue(.$p_beta)$q)

sum(sarkar_tophits$q_bonf <= 0.1, na.rm=T) #180
sum(sarkar_tophits$q_empirical <= 0.1, na.rm=T) #354
sum(sarkar_tophits$q_beta <= 0.1, na.rm=T) #279
```

### Power Comparison

```{r}
n_vqtl_tmm <- read_tsv(paste0("/project2/gilad/jpopp/ebQTL/results/static/vqtl_ipsc/pseudobulk_tmm/basic/IPSC/6pcs/matrixeqtl.cis_qtl_pairs.nhits.tsv"), col_names="n_egenes") #238
n_ebqtl_tmm <- read_tsv(paste0("/project2/gilad/jpopp/ebQTL/results/static/ebqtl_ipsc/pseudobulk_tmm/basic/IPSC/9pcs/matrixeqtl.cis_qtl_pairs.nhits.tsv"), col_names="n_egenes") #209
n_vqtl_scran <- read_tsv(paste0("/project2/gilad/jpopp/ebQTL/results/static/vqtl_ipsc/pseudobulk_scran/basic/IPSC/9pcs/matrixeqtl.cis_qtl_pairs.nhits.tsv"), col_names="n_egenes") #167
n_ebqtl_scran <- read_tsv(paste0("/project2/gilad/jpopp/ebQTL/results/static/ebqtl_ipsc/pseudobulk_scran/basic/IPSC/9pcs/matrixeqtl.cis_qtl_pairs.nhits.tsv"), col_names="n_egenes") #152
```


```{r}
ebqtl_tensorqtl <- read_tsv("/project2/gilad/jpopp/ebQTL/results/static/ebqtl_ipsc/pseudobulk_tmm/basic/IPSC/4pcs/tensorqtl_permutations.tsv")
sum(qvalue(ebqtl_tensorqtl$pval_perm)$q <= 0.1)
sum(qvalue(ebqtl_tensorqtl$pval_beta)$q <= 0.1)
```

```{r}
tibble(method=factor(c("vqtl_tmm", "vqtl_scran", "ebqtl_tmm", "ebqtl_scran", "vqtl_orig_bonf", "vqtl_orig_empirical"),
                       levels=c("ebqtl_scran", "ebqtl_tmm", "vqtl_scran", "vqtl_tmm", "vqtl_orig_bonf",
                                "vqtl_orig_empirical")),
       n=c(238, 167, 209, 152, 180, 354)) %>%
  ggplot(aes(x=method, y=n)) +
  geom_bar(stat="identity")
```

### Cross-study replication

First, lift over the variant locations to hg38

```{r}
sarkar_bed <- sarkar_tophits %>%
  select(var_chr, var_start, var_end) %>%
  mutate(var_start=var_start-1) %>%  #BED format
  write_tsv("/project2/gilad/jpopp/ebQTL/data/genotypes/sarkar.hits.hg19.bed", col_names=F)

sarkar_bed_hg38 <- read_tsv("/project2/gilad/jpopp/ebQTL/data/genotypes/sarkar.hits.hg38.liftover.bed", 
                            col_names=c("var_chr_hg38", "var_start_hg38", "var_end_hg38", "var_hg19", "lifted")) %>%
  select(-c(var_start_hg38, lifted)) %>%
  unique()

sarkar_tophits <- sarkar_tophits %>%
  unite(snp, var_start, var_end, sep="-") %>%
  unite(var_hg19, var_chr, snp, sep=":") %>%
  right_join(sarkar_bed_hg38, by="var_hg19") %>%
  unite(SNP, var_chr_hg38, var_end_hg38)
```

Next, convert the ENSG gene names to HGNC

```{r}
pull_ensg <- function(attr) {
  str_split(attr, "\"")[[1]][3]
}

pull_hgnc <- function(attr) {
  str_split(attr, "\"")[[1]][15]
}

gencode <- read_tsv("/project2/gilad/jpopp/ebQTL/data/gencode/gencode.hg38.filtered.gtf") %>%
  mutate(ensg=map_chr(attribute, pull_ensg)) %>%
  mutate(hgnc=map_chr(attribute, pull_hgnc))

gene_dict <- select(gencode, c(ensg, hgnc)) %>%
  filter(ensg %in% sarkar_all$gene) 

sarkar_tophits <- sarkar_tophits %>%
  left_join(gene_dict, by=c("gene"="ensg")) %>%
  drop_na() %>%
  select(-gene) %>% rename(gene=hgnc)
```

```{r output=FALSE}
ebqtl_tmm_all <- read_tsv("/project2/gilad/jpopp/ebQTL/results/static/ebqtl_ipsc/pseudobulk_tmm/basic/IPSC/9pcs/matrixeqtl.cis_qtl_pairs.all.tsv") %>%
  unite(gv, gene, SNP, sep="--", remove=F)
ebqtl_tmm_hits <- read_tsv("/project2/gilad/jpopp/ebQTL/results/static/ebqtl_ipsc/pseudobulk_tmm/basic/IPSC/9pcs/matrixeqtl.cis_qtl_pairs.tophits.tsv")

vqtl_tmm_all <- read_tsv("/project2/gilad/jpopp/ebQTL/results/static/vqtl_ipsc/pseudobulk_tmm/basic/IPSC/6pcs/matrixeqtl.cis_qtl_pairs.all.tsv") %>%
  unite(gv, gene, SNP, sep="--", remove=F)
vqtl_tmm_hits <- read_tsv("/project2/gilad/jpopp/ebQTL/results/static/vqtl_ipsc/pseudobulk_tmm/basic/IPSC/6pcs/matrixeqtl.cis_qtl_pairs.tophits.tsv")

vqtl_orig_bonf <- sarkar_tophits %>%
  filter(q_bonf <= 0.1) %>%
  unite(gv, gene, SNP, sep="--", remove=F)
```

```{r}
vqtl_tmm_all %>%
  filter(gv %in% vqtl_orig_bonf$gv) %>%
  pull(`p-value`) %>%
  hist

ebqtl_tmm_all %>%
  filter(gv %in% vqtl_orig_bonf$gv) %>%
  pull(`p-value`) %>%
  hist
```

Are the tests that we deem significant upon reanalysis, but they don't, totally ridiculous?

```{r}
new_hits <- vqtl_tmm_hits %>%
  filter(q <= 0.1) %>%
  filter(!gene %in% vqtl_orig_bonf$gene)

sarkar_tophits %>%
  filter(gene %in% new_hits$gene) %>% 
  pull(p_nominal) %>%
  hist
```

```{r}
overlap_tests <- intersect(ebqtl_tmm_all$gv, vqtl_tmm_all$gv)
vqtl_overlap <- filter(vqtl_tmm_all, gv %in% overlap_tests)  # 920454 of 990328
ebqtl_overlap <- filter(ebqtl_tmm_all, gv %in% overlap_tests) # 920454 of 1768879

vqtl_overlap <- vqtl_overlap %>%
  group_by(gene) %>% add_count() %>%
  mutate(bonf.p=if_else(`p-value`*n>1, 1, `p-value`*n)) %>%
  select(-n)

ebqtl_overlap <- ebqtl_overlap %>%
  group_by(gene) %>% add_count() %>%
  mutate(bonf.p=if_else(`p-value`*n>1, 1, `p-value`*n)) %>%
  select(-n)

vqtl_overlap_tophits <- vqtl_overlap %>%
  group_by(gene) %>%
  arrange(bonf.p) %>%
  filter(!duplicated(gene))
vqtl_overlap_tophits$q <- qvalue(vqtl_overlap_tophits$bonf.p)$qvalues

ebqtl_overlap_tophits <- ebqtl_overlap %>%
  group_by(gene) %>%
  arrange(bonf.p) %>%
  filter(!duplicated(gene))
ebqtl_overlap_tophits$q <- qvalue(ebqtl_overlap_tophits$bonf.p)$qvalues

vqtl_hits <- filter(vqtl_overlap_tophits, q<=0.1)$gv
hist(filter(ebqtl_overlap, gv %in% vqtl_hits)$`p-value`)

ebqtl_hits <- filter(ebqtl_overlap_tophits, q<=0.1)$gv
hist(qvalue(filter(vqtl_overlap, gv %in% ebqtl_hits)$`p-value`))
```

Excellent replication between the two studies
