---
title: "Static and Context-Specific eQTL Calling with mashR"
output: html_notebook
---


```{r}
library(mashr)
library(corrplot)
library(tidyverse)
library(rmeta)
```

## Quality Control
How many cells do we have per cell type and per individual?
```{r, fig.width=6.5, fig.height=20}
sample_summary <- read_tsv("/project2/gilad/jpopp/ebQTL/data/static/highpass_cellid_all/pseudobulk-scran/fastgxc/sample_summary_manual.tsv") %>%
  mutate(individual=factor(individual), type=factor(type))

sample_summary %>%
  arrange(n_cells) %>%
  ggplot(aes(x=individual, y=n_cells)) +
  geom_bar(stat="identity") +
  facet_wrap(vars(type), scales="free_y") +
  theme(axis.text.x = element_text(angle = 45, hjust=1))
```
How many cell types per individual?
```{r}
sample_summary %>%
  filter(!dropped) %>%
  dplyr::count(individual) %>%
  arrange(n) %>%
  ggplot(aes(x=individual, y=n)) +
  geom_bar(stat="identity") +
  theme(axis.text.x = element_text(angle = 45, hjust=1))
```


For each cell type, how many individuals have at least 5 cells? (This is the cutoff for having a pseudobulk sample)
```{r, fig.width=12, fig.height=6}
ct_summary <- read_tsv("/project2/gilad/jpopp/ebQTL/data/static/highpass_cellid_all/pseudobulk-scran/fastgxc/samples_per_celltype.tsv") %>%
  mutate(type=factor(type))

ct_summary  %>%
  ggplot(aes(x=type, y=n, fill=type)) +
  geom_bar(stat="identity") +
  ylab("n samples") +
  theme_classic(base_size=15) +
  theme(axis.text.x = element_text(angle = 45, hjust=1))
```
## Evaluating eQTL detection rates with and without mashR
How many eQTLs do we find with and without mashR?

### eQTL detection without mashR
```{r}
cell.types <- sample_summary  %>%
  mutate(type=as.character(type)) %>%
  filter(!dropped) %>%
  count(type) %>%
  filter(n > 25) %>%
  pull(type)
```

```{r}
pull_type <- function(s) {
  str_split(s, "/")[[1]][11]
}

prefix <- "/project2/gilad/jpopp/ebQTL/results/static/highpass_cellid_all/pseudobulk-scran/fastgxc/"
all_qtls <- vroom(paste0(prefix, cell.types, "/matrixeqtl.cis_qtl_pairs.tophits.tsv"), id="path", col_select=c(SNP, gene, bonf.p, q, path)) %>%
  mutate(type=sapply(path, pull_type), .keep="unused")
```

```{r}
qtls_per_type <- all_qtls %>%
  unite(gv, SNP, gene, sep="--") %>%
  select(gv, q, type) %>%
  group_by(type) %>%
  mutate(hit=(q <= 0.05)) %>% 
  summarize(n=sum(hit))

ggplot(qtls_per_type, aes(x=type, y=n)) +
  geom_segment( aes(x=type, xend=type, y=0, yend=n), color="skyblue") +
  geom_point( color="blue", size=8, alpha=0.6) +
  theme_light(base_size=25) +
  coord_flip() +
  theme(
    panel.grid.major.y = element_blank(),
    panel.border = element_blank(),
    axis.ticks.y = element_blank()
  ) +
  ylab("# Genes with eQTL") + xlab("FCA-Annotated Cell Type")
```

What's driving the difference in detection rates?
```{r}
sample_summary  %>%
  mutate(type=as.character(type)) %>%
  filter(!dropped) %>%
  count(type) %>% rename(n_inds=n) %>%
  inner_join(qtls_per_type, by="type") %>%
  rename(n_qtl=n) %>%
  ggplot(aes(x=n_inds, y=n_qtl)) +
  geom_point(size=4) +
  theme_classic(base_size=20) +
  theme(legend.position = "none")
```

### eQTL detection with mashR
```{r}
m <- readRDS("/project2/gilad/jpopp/ebQTL/results/static/highpass_cellid_all/pseudobulk-scran/fastgxc/mashr.tophits.rds")
```

```{r}
n_qtl <- colSums(m$result$lfsr <= 0.05)

qtls_per_type_mashr <- tibble(type=names(n_qtl), n_eqtl=n_qtl)  %>% 
  arrange(n_qtl)

ggplot(qtls_per_type_mashr, aes(x=type, y=n_eqtl)) +
  geom_segment( aes(x=type, xend=type, y=0, yend=n_eqtl), color="skyblue") +
  geom_point( color="blue", size=8, alpha=0.6) +
  theme_light(base_size=25) +
  coord_flip() +
  theme(
    panel.grid.major.y = element_blank(),
    panel.border = element_blank(),
    axis.ticks.y = element_blank()
  ) +
  ylab("# Genes with eQTL") + xlab("FCA-Annotated Cell Type")
```

## Understanding Data-Driven Regulatory Patterns
```{r}
get_pairwise_sharing(m) %>%
  corrplot(is.corr=F, order='hclust', tl.pos='lt')
```

How many eQTLs do we find?
```{r}
sum(get_n_significant_conditions(m) >= 1)
```


```{r}
tibble(n_contexts=get_n_significant_conditions(m)) %>%
  filter(n_contexts >= 1) %>%
  dplyr::count(n_contexts, name="n_eqtl") %>%
  mutate(n_contexts=factor(n_contexts, levels=seq(1, ncol(m$result$lfsr)))) %>%
  ggplot(aes(x=n_contexts, y=n_eqtl)) +
  geom_bar(stat="identity") +
  theme_classic(base_size=15)
```

```{r}
n_qtl <- colSums(m$result$lfsr <= 0.05)

qtl_by_type <- tibble(type=names(n_qtl), n_eqtl=n_qtl)  %>% 
  arrange(n_qtl)

ggplot(qtl_by_type, aes(x=type, y=n_eqtl, fill=type)) +
  geom_bar(stat="identity") +
  theme_classic(base_size=15) +
  theme(axis.text.x = element_text(angle=45, hjust=1), legend.position="none")

ggplot(qtl_by_type, aes(x=type, y=n_eqtl)) +
  geom_segment( aes(x=type, xend=type, y=0, yend=n_eqtl), color="skyblue") +
  geom_point( color="blue", size=8, alpha=0.6) +
  theme_light(base_size=25) +
  coord_flip() +
  theme(
    panel.grid.major.y = element_blank(),
    panel.border = element_blank(),
    axis.ticks.y = element_blank()
  ) +
  ylab("# Genes with eQTL") + xlab("FCA-Annotated Cell Type")
```

```{r}
tibble(n_contexts=get_n_significant_conditions(m)) %>%
  filter(n_contexts >= 1) %>%
  nrow(.)
```

Now, of the eQTLs that are specific to one context, which context is it?
```{r}
m$result$lfsr %>%
  as_tibble(rownames="gv") %>%
  pivot_longer(-gv, names_to="type", values_to="lfsr") %>%
  filter(lfsr<=0.05) %>%
  group_by(gv) %>%
  dplyr::select(-lfsr) %>%
  nest() %>%
  mutate(n_contexts=sapply(data, nrow)) %>%
  filter(n_contexts==1) %>%
  unnest(data) %>%
  mutate(type=factor(type)) %>%
  group_by(type) %>%
  dplyr::count() %>%
  ggplot(aes(x=type, y=n, fill=type)) +
  geom_bar(stat="identity") +
  theme_classic() +
  ylab("# Singleton eQTLs") +
  theme(axis.text.x = element_text(angle = 45, hjust=1))
```
There's a ton for erythroblasts - needs follow-up. Maybe this is because those are not really erythroblasts but they're from the early ectoderm cluster where we have tons of cells.

Do we get more eQTLs from the contexts where we have more individuals?
```{r}

```



### What are the strongest patterns?
```{r, fig.width=6, fig.height=8}
as_tibble(get_estimated_pi(m), rownames="component") %>%
  mutate(component=factor(component, levels=names(get_estimated_pi(m)))) %>%
  ggplot(aes(x=component, y=value)) +
  geom_bar(stat="identity") +
  scale_x_discrete(limits=rev) +
  xlab("Regulatory Pattern") +
  ylab("Weight") +
  coord_flip() +
  theme_classic(base_size = 20)
  #theme(axis.text.x = element_text(angle = 45, hjust=1))
```

Show more interesting covariance matrices
```{r}
germ_layer <- read_csv("/project2/gilad/katie/ebQTL/FetalReferenceCellTypes_GermLayerAnnotations.csv", col_select = -c(Notes)) %>%
  rename(celltype=`cell type`, layer=`Germ layer`) %>%
  group_by(celltype) %>% add_count() %>%
  mutate(layer=if_else(n > 1, "Ambiguous", layer)) %>%
  dplyr::select(celltype, layer) %>%
  distinct() %>%
  mutate(layer=factor(layer, levels=c("Ambiguous", "Ectoderm", "Neural crest", "Endoderm", "Mesoderm"))) %>%
  mutate(celltype=str_replace_all(celltype, c(" "="-", "_"="-")))

germlayer_levels <- c("Ambiguous", "Ectoderm", "Neural crest", "Endoderm", "Mesoderm")
germlayer_colors <- c("#D5D5D5", "#0076BA", "#006C65", "#FEAE00", "#A62228")
```

The main component that mashr uses draws a clean split down germ layer lines
```{r}
m$fitted_g$Ulist$ED_tPCA %>%
  corrplot(is.corr=F)

pcs <- m$fitted_g$Ulist$ED_tPCA %>%
  princomp %>%
  .$loadings %>%
  .[,1] %>%
  as_tibble(rownames="celltype") %>%
  left_join(germ_layer, by="celltype")
 
pcs %>%
  ggplot(aes(x=celltype, y=value, fill=layer)) +
  geom_bar(stat="identity") +
  scale_fill_manual(values=germlayer_colors) +
  coord_flip() +
  theme_classic()
```

```{r}
m$fitted_g$Ulist$ED_PCA_1 %>%
  corrplot(is.corr=F, tl.pos='l')

pc1 <- m$fitted_g$Ulist$ED_PCA_1 %>%
  princomp %>%
  .$loadings %>%
  .[,1] %>%
  as_tibble(rownames="celltype") %>%
  left_join(germ_layer, by="celltype")
 
pc1 %>%
  ggplot(aes(x=celltype, y=value, fill=layer)) +
  geom_bar(stat="identity") +
  scale_fill_manual(values=germlayer_colors) +
  coord_flip() +
  theme_classic()

m$fitted_g$Ulist$ED_PCA_2 %>%
  corrplot(is.corr=F, tl.pos='l')

pc2 <- m$fitted_g$Ulist$ED_PCA_2 %>%
  princomp %>%
  .$loadings %>%
  .[,1] %>%
  as_tibble(rownames="celltype") %>%
  left_join(germ_layer, by="celltype")
 
pc2 %>%
  ggplot(aes(x=celltype, y=value, fill=layer)) +
  geom_bar(stat="identity") +
  scale_fill_manual(values=germlayer_colors) +
  coord_flip() +
  theme_classic()
```


## Examining Context-Specific eQTLs
Find any significant eQTLs whose effects are best explained by those signals, starting with component 1 (which is highest in IPSC, lower elsewhere)
```{r}
tests <- rownames(get_lfsr(m))
sighits <- get_significant_results(m)
ident <- as_tibble(m$posterior_weights) %>%
  rowid_to_column() %>%
  filter(rowid %in% sighits) %>%
  pivot_longer(-rowid, names_to="component", values_to="weight") %>%
  group_by(rowid) %>%
  slice_max(weight) %>%
  mutate(component=str_extract(component, "[^.]+")) %>%
  filter(component=="ED_tPCA") %>%
  pull(rowid)

celltype.colors <- germ_layer %>%
  left_join(tibble(layer=germlayer_levels, color=germlayer_colors), by="layer") %>%
  arrange(celltype) %>%
  filter(celltype %in% colnames(m$result$lfsr)) %>%
  pull(color)

mash_plot_meta(m, ident[1], colors=meta.colors(box=celltype.colors))
mash_plot_meta(m, ident[2], colors=meta.colors(box=celltype.colors))
mash_plot_meta(m, ident[3], colors=meta.colors(box=celltype.colors))
mash_plot_meta(m, ident[4], colors=meta.colors(box=celltype.colors))
mash_plot_meta(m, ident[5], colors=meta.colors(box=celltype.colors))
mash_plot_meta(m, ident[6], colors=meta.colors(box=celltype.colors))
mash_plot_meta(m, ident[7], colors=meta.colors(box=celltype.colors))
```
## Shared eQTL Effects
Next, we want to see how many eQTL effects are shared across all of our contexts. 

We trained a (univariate) `ash` model on the same set of tests used for our `mash` analysis.
```{r}
ash_tophits <- readRDS("/project2/gilad/jpopp/ebQTL/results/static/highpass_cellid_all/pseudobulk-scran/fastgxc/ashr.tophits.rds")
sum(ash_tophits$result$lfsr <= 0.05)
```

So, this analysis reveals 138 genes with a shared eQTL. Let's look at the histogram
```{r}
as_tibble(ash_tophits$result) %>%
  mutate(t=betahat/sebetahat, signif=(lfsr <= 0.05)) %>%
  ggplot(aes(x=t, fill=signif)) +
  geom_histogram(bins=100) +
  theme_classic()
```

## Comparing shared and specific regulatory effects
How many of our context-specific regulatory effects overlap with shared effects?

This would suggest that at these loci the effect *size* varies, though the effect is always present.
```{r}
sp_eqtls <- tibble(gv=names(get_significant_results(m))) %>%
  separate(gv, into=c("gene", "snp"), sep="--", remove=F)

sh_eqtls <- tibble(gv=names(ash_tophits$data$x), lfsr=get_lfsr(ash_tophits)) %>%
  filter(lfsr <= 0.05) %>%
  dplyr::select(gv) %>%
  separate(gv, into=c("gene", "snp"), sep="--", remove=F)

length(intersect(sh_eqtls$gv, sp_eqtls$gv))
```

Next, I'd like to see how many of these were detected in GTEx. First, I will load all GTEx lead variants.
```{r}

```

